{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-15T09:28:54.601864Z",
     "start_time": "2018-05-15T09:28:54.554930Z"
    }
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "version = \"1.1.0\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_py_file_path(py_file_path):\n",
    "\n",
    "    py_file_name == py_file_path.split(\"/\")[-1]\n",
    "\n",
    "    py_file_name_without_py = py_file_name[: -len(\".py\")]\n",
    "\n",
    "    if py_file_name_without_py.isupper():\n",
    "\n",
    "        importable_names = [py_file_name_without_py]\n",
    "\n",
    "    else:\n",
    "\n",
    "        importable_names = []\n",
    "\n",
    "        with open(py_file_path) as py_file:\n",
    "\n",
    "            for line in py_file.readlines():\n",
    "\n",
    "                for prefix in (\"def \", \"class \"):\n",
    "                        \n",
    "                    if line.startswith(prefix):\n",
    "                        \n",
    "                        if '(' in line:\n",
    "                            \n",
    "                            separator = '('\n",
    "                        \n",
    "                        elif ':' in line:\n",
    "                        \n",
    "                            separator = ':'\n",
    "\n",
    "                        importable_names.append(\n",
    "                            line.strip().split(separator)[0][len(prefix) :]\n",
    "                        )\n",
    "                        \n",
    "    _names = [name for name in importable_names if name.startswith('_')]\n",
    "    \n",
    "    if len(_names):\n",
    "        \n",
    "        print(\"\\t\" + \"\\n\\t\".join(_names))\n",
    "    \n",
    "    importable_names = set(importable_names) - set(_names)\n",
    "            \n",
    "    if len(importable_names) != 1:\n",
    "\n",
    "        raise ValueError(\"N importable names ({}) != 1\".format(len(importable_names)))\n",
    "\n",
    "    importable_name = importable_names.pop()\n",
    "\n",
    "    if importable_name != py_file_name_without_py:\n",
    "\n",
    "        raise ValueError(\n",
    "            \"{}(.py) != {}\".format(py_file_name_without_py, importable_name)\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ALMOST_ZERO.py\n",
      "\n",
      "CODON_AMINO_ACID.py\n",
      "\n",
      "COLORS.py\n",
      "\n",
      "DATA_DIRECTORY_PATH.py\n",
      "\n",
      "FeatureHDF5.py\n",
      "\t_FeatureDescription\n",
      "\n",
      "GPSMap.py\n",
      "\n",
      "Genome.py\n",
      "\n",
      "NONE_STRS.py\n",
      "\n",
      "RANDOM_SEED.py\n",
      "\n",
      "VCF_ANN_FIELDS.py\n",
      "\n",
      "VCF_COLUMNS.py\n",
      "\n",
      "VariantHDF5.py\n",
      "\t_VariantDescription\n",
      "\n",
      "__init__.py\n",
      "\n",
      "__pycache__\n",
      "\n",
      "add_conda_to_path.py\n",
      "\n",
      "anneal_gps_map_node_and_element_positions.py\n",
      "\n",
      "apply_function_on_2_1d_arrays.py\n",
      "\n",
      "apply_function_on_2_2d_arrays_slices.py\n",
      "\n",
      "apply_function_on_2_dfs_slices.py\n",
      "\n",
      "call_function_with_multiporcess.py\n",
      "\n",
      "cast_series_to_builtin.py\n",
      "\n",
      "cast_str_to_builtin.py\n",
      "\n",
      "check_gps_map_node_x_element.py\n",
      "\n",
      "check_nd_array_for_bad.py\n",
      "\n",
      "clean_and_write_df_to_tsv.py\n",
      "\n",
      "clip_nd_array_by_standard_deviation.py\n",
      "\n",
      "cluster_2d_array.py\n",
      "\n",
      "cluster_clustering_x_element_and_compute_ccc.py\n",
      "\n",
      "compute_1d_array_context.py\n",
      "\n",
      "compute_1d_array_entropy.py\n",
      "\n",
      "compute_1d_arrays_bandwidths.py\n",
      "\n",
      "compute_coclustering_fraction_from_clustering_x_element.py\n",
      "\n",
      "compute_context_indices_from_pdf.py\n",
      "\n",
      "compute_correlation_distance_between_2_1d_arrays.py\n",
      "\n",
      "compute_empirical_p_value.py\n",
      "\n",
      "compute_empirical_p_values_and_fdrs.py\n",
      "\n",
      "compute_information_coefficient_between_2_1d_arrays.py\n",
      "\n",
      "compute_information_distance_between_2_1d_arrays.py\n",
      "\n",
      "compute_joint_probability.py\n",
      "\n",
      "compute_kullback_leibler_divergence_between_2_pdfs.py\n",
      "\n",
      "compute_matrix_norm.py\n",
      "\n",
      "compute_nd_array_margin_of_error.py\n",
      "\n",
      "compute_posterior_probability.py\n",
      "\n",
      "correlate_2_1d_arrays.py\n",
      "\n",
      "count_gene_impacts_from_variant_dicts.py\n",
      "\n",
      "create_gitkeep.py\n",
      "\n",
      "download_and_parse_geo.py\n",
      "\n",
      "download_url.py\n",
      "\n",
      "drop_df_slice.py\n",
      "\n",
      "drop_df_slice_greedily.py\n",
      "\n",
      "echo_or_print_str.py\n",
      "\n",
      "establish_path.py\n",
      "\n",
      "estimate_kernel_density.py\n",
      "\n",
      "exit_.py\n",
      "\n",
      "fit_skew_t_pdf_on_1d_array.py\n",
      "\n",
      "fit_skew_t_pdf_on_each_df_row.py\n",
      "\t_fit_skew_t_pdf_on_each_df_row\n",
      "\n",
      "flatten_nested_iterable.py\n",
      "\n",
      "get_absolute_path.py\n",
      "\n",
      "get_colormap_colors.py\n",
      "\n",
      "get_colorscale_for_data.py\n",
      "\n",
      "get_conda_environments.py\n",
      "\n",
      "get_conda_prefix.py\n",
      "\n",
      "get_gff3_attribute.py\n",
      "\n",
      "get_git_versions.py\n",
      "\n",
      "get_installed_pip_libraries.py\n",
      "\n",
      "get_intersections_between_2_1d_arrays.py\n",
      "\n",
      "get_machine.py\n",
      "\n",
      "get_name_within_function.py\n",
      "\n",
      "get_sequence_from_fasta_gz.py\n",
      "\n",
      "get_shell_environment.py\n",
      "\n",
      "get_target_grid_indices.py\n",
      "\n",
      "get_triangulation_edges_from_point_x_dimension.py\n",
      "\n",
      "get_variant_start_and_end_positions.py\n",
      "\n",
      "get_vcf_allelic_frequencies.py\n",
      "\n",
      "get_vcf_genotype.py\n",
      "\n",
      "get_vcf_info.py\n",
      "\n",
      "get_vcf_info_ann.py\n",
      "\n",
      "get_vcf_population_allelic_frequencies.py\n",
      "\n",
      "get_vcf_sample_format.py\n",
      "\n",
      "get_vcf_variants_by_region.py\n",
      "\n",
      "group_and_apply_function_on_each_group_in_iterable.py\n",
      "\n",
      "gsea.py\n",
      "\n",
      "hierarchical_consensus_cluster.py\n",
      "\n",
      "hierarchical_consensus_cluster_with_ks.py\n",
      "\n",
      "ignore_bad_and_compute_euclidean_distance_between_2_1d_arrays.py\n",
      "\n",
      "index_gff3_df_by_name.py\n",
      "\n",
      "infer.py\n",
      "\n",
      "infer_assuming_independence.py\n",
      "\n",
      "initialize_logger.py\n",
      "\n",
      "install_and_activate_conda.py\n",
      "\n",
      "install_python_libraries.py\n",
      "\n",
      "is_in_git_repository.py\n",
      "\n",
      "is_program.py\n",
      "\n",
      "is_sorted_nd_array.py\n",
      "\n",
      "is_str_version.py\n",
      "\n",
      "is_valid_conda_directory_path.py\n",
      "\n",
      "is_valid_vcf_gz.py\n",
      "\n",
      "log_and_return_response.py\n",
      "\n",
      "log_nd_array.py\n",
      "\n",
      "make_binary_df_from_categorical_series.py\n",
      "\n",
      "make_categorical_colors.py\n",
      "\n",
      "make_colorscale_from_colors.py\n",
      "\n",
      "make_context_matrix.py\n",
      "\t_make_context_matrix\n",
      "\n",
      "make_coordinates_for_reflection.py\n",
      "\n",
      "make_gps_map_element_x_dimension.py\n",
      "\n",
      "make_gps_map_grid_values_and_categorical_labels.py\n",
      "\n",
      "make_match_panel.py\n",
      "\n",
      "make_match_panel_annotations.py\n",
      "\n",
      "make_match_panels.py\n",
      "\n",
      "make_mesh_grid_coordinates_per_axis.py\n",
      "\n",
      "make_object_int_and_int_object_for_iterable.py\n",
      "\n",
      "make_summary_match_panel.py\n",
      "\n",
      "make_variant_dict_consistent.py\n",
      "\n",
      "make_variant_dict_from_vcf_row.py\n",
      "\n",
      "match_randomly_sampled_target_and_data_to_compute_margin_of_errors.py\n",
      "\n",
      "match_target_and_data.py\n",
      "\n",
      "match_target_and_data_and_compute_statistics.py\n",
      "\n",
      "merge_dicts_with_function.py\n",
      "\n",
      "mf_by_multiple_V_and_H.py\n",
      "\n",
      "mf_by_multiplicative_update.py\n",
      "\n",
      "mf_consensus_cluster.py\n",
      "\n",
      "mf_consensus_cluster_with_ks.py\n",
      "\n",
      "nmf_by_sklearn.py\n",
      "\n",
      "normalize_cell_line_names.py\n",
      "\n",
      "normalize_contig.py\n",
      "\n",
      "normalize_file_name.py\n",
      "\n",
      "normalize_git_url.py\n",
      "\n",
      "normalize_nd_array.py\n",
      "\t_normalize_nd_array\n",
      "\n",
      "permute_target_and_match_target_and_data.py\n",
      "\n",
      "plot_and_save.py\n",
      "\n",
      "plot_bayesian_nomogram.py\n",
      "\n",
      "plot_bubble_map.py\n",
      "\n",
      "plot_context.py\n",
      "\n",
      "plot_gps_map.py\n",
      "\n",
      "plot_gsea_mountain.py\n",
      "\n",
      "plot_heat_map.py\n",
      "\n",
      "plot_histogram.py\n",
      "\n",
      "plot_scatter_and_annotate.py\n",
      "\n",
      "process_df_for_plotting.py\n",
      "\n",
      "process_feature_x_sample.py\n",
      "\n",
      "process_match_panel_target_or_data_for_plotting.py\n",
      "\n",
      "read_firebrowse_copynumber_gistic2.py\n",
      "\n",
      "read_firebrowse_correlate_copynumber_vs_mrnaseq.py\n",
      "\n",
      "read_firebrowse_mutsignozzlereport2cv.py\n",
      "\n",
      "read_gff3.py\n",
      "\n",
      "read_gmt.py\n",
      "\n",
      "read_gps_map.py\n",
      "\n",
      "read_json.py\n",
      "\n",
      "read_matrix_market.py\n",
      "\n",
      "read_vcf_gz_and_make_vcf_dict.py\n",
      "\n",
      "read_where_and_map_column_name_on_hdf5_table.py\n",
      "\n",
      "reduce_point_x_dimension_dimension.py\n",
      "\n",
      "rescale_x_y_coordiantes_in_polar_coordiante.py\n",
      "\n",
      "reverse_complement_dna_sequence.py\n",
      "\n",
      "reverse_transcribe_rna_sequence.py\n",
      "\n",
      "run_command.py\n",
      "\n",
      "sample_from_each_series_value.py\n",
      "\n",
      "select_and_group_feature_x_tcga_sample_by_sample_type.py\n",
      "\n",
      "select_gene_symbol.py\n",
      "\n",
      "select_series_indices.py\n",
      "\n",
      "shuffle_each_2d_array_slice.py\n",
      "\n",
      "single_sample_gsea.py\n",
      "\n",
      "single_sample_gseas.py\n",
      "\t_single_sample_gseas\n",
      "\n",
      "skip_quote_and_split_str.py\n",
      "\n",
      "solve_ax_equal_b.py\n",
      "\n",
      "solve_for_H.py\n",
      "\n",
      "split_codons.py\n",
      "\n",
      "split_df.py\n",
      "\n",
      "summarize_feature_x_sample.py\n",
      "\n",
      "title_str.py\n",
      "\n",
      "train_and_classify.py\n",
      "\n",
      "train_and_regress.py\n",
      "\n",
      "transcribe_dna_sequence.py\n",
      "\n",
      "translate_nucleotide_sequence.py\n",
      "\n",
      "untitle_str.py\n",
      "\n",
      "update_H_by_multiplicative_update.py\n",
      "\n",
      "update_W_by_multiplicative_update.py\n",
      "\n",
      "update_variant_dict.py\n",
      "\n",
      "write_gps_map.py\n",
      "\n",
      "write_json.py\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import_lines = []\n",
    "\n",
    "for py_file_name in sorted(os.listdir(\"../ccal\")):\n",
    " \n",
    "    print(py_file_name)\n",
    "\n",
    "    if py_file_name != \"__init__.py\" and py_file_name.endswith(\".py\"):\n",
    "\n",
    "        check_py_file_path(\"../ccal/{}\".format(py_file_name))\n",
    "\n",
    "        import_lines.append(\"from .{0} import {0}\".format(py_file_name[: -len(\".py\")]))\n",
    "        \n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../ccal/__init__.py\", mode=\"w\") as init_file:\n",
    "\n",
    "    init_file.write(\n",
    "        \"\\n\".join(\n",
    "            (\n",
    "                \"from os.path import abspath\",\n",
    "                'VERSION = \"{}\"'.format(version),\n",
    "                'print(\"CCAL version {} @ {}\".format(VERSION, abspath(__file__)))',\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "\n",
    "    init_file.write(\"\\n\")\n",
    "\n",
    "    init_file.writelines(\"\\n\".join(import_lines))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCAL version 1.1.0 @ /home/kwatme/github/ccal/ccal/__init__.py\n"
     ]
    },
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'ccal.plot_points'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-4156f152e9b4>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minsert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"..\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mccal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/github/ccal/ccal/__init__.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mDATA_DIRECTORY_PATH\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mDATA_DIRECTORY_PATH\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mFeatureHDF5\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mFeatureHDF5\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mGPSMap\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mGPSMap\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mGenome\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mGenome\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mNONE_STRS\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mNONE_STRS\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/github/ccal/ccal/GPSMap.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mscipy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspatial\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdistance\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpdist\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msquareform\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m from .anneal_gps_map_node_and_element_positions import (\n\u001b[0m\u001b[1;32m      7\u001b[0m     \u001b[0manneal_gps_map_node_and_element_positions\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m )\n",
      "\u001b[0;32m~/github/ccal/ccal/anneal_gps_map_node_and_element_positions.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mapply_function_on_2_2d_arrays_slices\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mapply_function_on_2_2d_arrays_slices\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mplot_points\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mplot_points\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'ccal.plot_points'"
     ]
    }
   ],
   "source": [
    "sys.path.insert(0, \"..\")\n",
    "\n",
    "import ccal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(dir(ccal)))\n",
    "\n",
    "dir(ccal)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
