{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-15T09:28:54.601864Z",
     "start_time": "2018-05-15T09:28:54.554930Z"
    }
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "version = \"0.9.1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_py_file_path(py_file_path):\n",
    "\n",
    "    py_file_name == py_file_path.split(\"/\")[-1]\n",
    "\n",
    "    py_file_name_without_py = py_file_name[: -len(\".py\")]\n",
    "\n",
    "    if py_file_name_without_py.isupper():\n",
    "\n",
    "        importable_names = [py_file_name_without_py]\n",
    "\n",
    "    else:\n",
    "\n",
    "        importable_names = []\n",
    "\n",
    "        with open(py_file_path) as py_file:\n",
    "\n",
    "            for line in py_file.readlines():\n",
    "\n",
    "                for prefix, separator in ((\"def \", \"(\"), (\"class \", \":\")):\n",
    "\n",
    "                    if line.startswith(prefix):\n",
    "\n",
    "                        importable_names.append(\n",
    "                            line.strip().split(separator)[0][len(prefix) :]\n",
    "                        )\n",
    "\n",
    "    if len(importable_names) != 1:\n",
    "\n",
    "        raise ValueError(\"N importable names ({}) != 1\".format(len(importable_names)))\n",
    "\n",
    "    importable_name = importable_names[0]\n",
    "\n",
    "    if importable_name != py_file_name_without_py:\n",
    "\n",
    "        raise ValueError(\n",
    "            \"{}(.py) != {}\".format(py_file_name_without_py, importable_name)\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BAD_VARIANT_IDS.py\n",
      "CODON_TO_AMINO_ACID.py\n",
      "COLOR_CATEGORICAL.py\n",
      "COLOR_RUBY_EMERALD.py\n",
      "COLOR_WHITE_BLACK.py\n",
      "COLOR_WHITE_BROWN.py\n",
      "DATA_DIRECTORY_PATH.py\n",
      "FeatureHDF5.py\n",
      "GPSMap.py\n",
      "Genome.py\n",
      "VARIANT_CLASSIFICATION_MUTSIG_EFFECT.py\n",
      "VARIANT_EFFECTS.py\n",
      "VCF_ANN_FIELDS.py\n",
      "VCF_COLUMNS.py\n",
      "VariantHDF5.py\n",
      "__init__.py\n",
      "__pycache__\n",
      "_anneal_node_and_element_positions.py\n",
      "_check_fastq_gzs.py\n",
      "_check_node_x_element.py\n",
      "_check_w_or_h.py\n",
      "_cluster_clustering_x_element_and_compute_ccc.py\n",
      "_compute_context_indices.py\n",
      "_compute_norm.py\n",
      "_count.py\n",
      "_describe_vcf_df.py\n",
      "_fit_skew_t_pdfs.py\n",
      "_get_coclustering_portion.py\n",
      "_get_target_grid_indices.py\n",
      "_get_triangulation_edges.py\n",
      "_gzip_compress.py\n",
      "_identify_what_to_count.py\n",
      "_ignore_bad_and_compute_euclidean_distance_between_2_1d_arrays.py\n",
      "_make_annotations.py\n",
      "_make_clean_vcf_df.py\n",
      "_make_context_matrix.py\n",
      "_make_element_x_dimension.py\n",
      "_make_grid_values_and_categorical_labels.py\n",
      "_make_variant_dict_consistent.py\n",
      "_match.py\n",
      "_match_randomly_sampled_target_and_data_to_compute_margin_of_errors.py\n",
      "_match_target_and_data.py\n",
      "_normalize_nd_array.py\n",
      "_permute_target_and_match_target_and_data.py\n",
      "_plot_2d.py\n",
      "_plot_gps_map.py\n",
      "_plot_mountain.py\n",
      "_print_and_run_command.py\n",
      "_process_target_or_data_for_plotting.py\n",
      "_single_sample_gseas.py\n",
      "_update_H_by_multiplicative_update.py\n",
      "_update_W_by_multiplicative_update.py\n",
      "add_conda_to_path.py\n",
      "align_fastq_gzs_using_bwa_mem.py\n",
      "align_fastq_gzs_using_hisat2.py\n",
      "annotate_vcf_gz_using_bcftools_annotate.py\n",
      "annotate_vcf_gz_using_snpeff.py\n",
      "apply_function_on_2_1d_arrays.py\n",
      "apply_function_on_2_2d_arrays_slices.py\n",
      "bgzip_and_tabix.py\n",
      "cast_series_to_builtins.py\n",
      "cast_str_to_builtins.py\n",
      "check_bam_using_samtools_flagstat.py\n",
      "check_fastq_gzs_using_fastqc.py\n",
      "check_nd_array_for_bad.py\n",
      "clean_and_write_df_to_tsv.py\n",
      "clean_git_url.py\n",
      "clean_name.py\n",
      "clean_path.py\n",
      "clip_nd_array_by_standard_deviation.py\n",
      "cluster_2d_array_slices.py\n",
      "compute_bandwidths.py\n",
      "compute_context.py\n",
      "compute_correlation_distance.py\n",
      "compute_empirical_p_value.py\n",
      "compute_empirical_p_values_and_fdrs.py\n",
      "compute_entropy.py\n",
      "compute_information_coefficient.py\n",
      "compute_information_distance.py\n",
      "compute_joint_probability.py\n",
      "compute_kullback_leibler_divergence.py\n",
      "compute_mutational_signature_enrichment.py\n",
      "compute_nd_array_margin_of_error.py\n",
      "compute_posterior_probability.py\n",
      "concatenate_vcf_gzs_using_bcftools_concat.py\n",
      "conda_is_installed.py\n",
      "copy_path.py\n",
      "correlate.py\n",
      "count_gene_impacts_from_variant_dicts.py\n",
      "count_transcripts_using_kallisto_quant.py\n",
      "count_vcf_gz_rows.py\n",
      "create_gitkeep.py\n",
      "cross_validate.py\n",
      "download.py\n",
      "download_and_parse_geo_data.py\n",
      "download_clinvar_vcf_gz.py\n",
      "drop_df_slice.py\n",
      "drop_df_slice_greedily.py\n",
      "dump_gps_map.py\n",
      "echo_or_print.py\n",
      "establish_fai_index.py\n",
      "establish_path.py\n",
      "estimate_kernel_density.py\n",
      "exit_.py\n",
      "faidx_fasta.py\n",
      "filter_vcf_gz_using_bcftools_view.py\n",
      "fit_skew_t_pdf.py\n",
      "fit_skew_t_pdfs.py\n",
      "flatten_nested_iterable.py\n",
      "get_1d_array_unique_objects_in_order.py\n",
      "get_allelic_frequencies.py\n",
      "get_chromosome_size_from_fasta_gz.py\n",
      "get_colormap_colors.py\n",
      "get_conda_environments.py\n",
      "get_conda_prefix.py\n",
      "get_function_name.py\n",
      "get_genotype.py\n",
      "get_gff3_attribute.py\n",
      "get_git_versions.py\n",
      "get_installed_pip_libraries.py\n",
      "get_intersections_between_2_1d_arrays.py\n",
      "get_machine.py\n",
      "get_maf_variant_classification.py\n",
      "get_now.py\n",
      "get_object_reference.py\n",
      "get_open_port.py\n",
      "get_population_allelic_frequencies.py\n",
      "get_sequence_from_fasta_gz.py\n",
      "get_shell_environment.py\n",
      "get_unique_iterable_objects_in_order.py\n",
      "get_variant_start_and_end_positions.py\n",
      "get_variant_type.py\n",
      "get_variants_from_bam_using_freebayes.py\n",
      "get_variants_from_bam_using_freebayes_and_multiprocess.py\n",
      "get_variants_from_bam_using_strelka.py\n",
      "get_variants_from_vcf_gz.py\n",
      "get_vcf_info.py\n",
      "get_vcf_info_ann.py\n",
      "get_vcf_sample_format.py\n",
      "get_volume_name.py\n",
      "group_and_apply_function_on_each_group_in_iterable.py\n",
      "group_iterable.py\n",
      "gsea.py\n",
      "gzip_compress_file.py\n",
      "gzip_decompress_and_bgzip_compress_file.py\n",
      "gzip_decompress_file.py\n",
      "have_program.py\n",
      "hierarchical_consensus_cluster.py\n",
      "hierarchical_consensus_cluster_with_ks.py\n",
      "in_git_repository.py\n",
      "index_bam_using_samtools_index.py\n",
      "index_gff3_df_by_name.py\n",
      "infer.py\n",
      "infer_assuming_independence.py\n",
      "initialize_logger.py\n",
      "install_and_activate_conda.py\n",
      "install_python_libraries.py\n",
      "is_inframe.py\n",
      "is_valid_vcf_gz.py\n",
      "load_gps_map.py\n",
      "log_and_return_response.py\n",
      "log_nd_array.py\n",
      "make_case_annotations.py\n",
      "make_categorical_colors.py\n",
      "make_colorscale.py\n",
      "make_colorscale_from_colors.py\n",
      "make_comparison_panel.py\n",
      "make_context_matrix.py\n",
      "make_coordinates_for_reflection.py\n",
      "make_file_name_from_str.py\n",
      "make_maf_from_vcf.py\n",
      "make_match_panel.py\n",
      "make_match_panels.py\n",
      "make_membership_df_from_categorical_series.py\n",
      "make_mesh_grid_coordinates_per_axis.py\n",
      "make_object_int_mapping.py\n",
      "make_random_color.py\n",
      "make_reference_genome.py\n",
      "make_summary_match_panel.py\n",
      "make_volume_dict.py\n",
      "map_cell_line_names.py\n",
      "mark_duplicates_in_bam_using_picard_markduplicates.py\n",
      "mds.py\n",
      "merge_dicts_with_callable.py\n",
      "mount_volume.py\n",
      "multiprocess.py\n",
      "nd_array_is_sorted.py\n",
      "nmf_by_multiple_V_and_H.py\n",
      "nmf_by_multiplicative_update.py\n",
      "nmf_by_sklearn.py\n",
      "nmf_consensus_cluster.py\n",
      "nmf_consensus_cluster_with_ks.py\n",
      "normalize_contig.py\n",
      "normalize_df.py\n",
      "normalize_nd_array.py\n",
      "parse_vcf_row_and_make_variant_dict.py\n",
      "plot_and_save.py\n",
      "plot_bar.py\n",
      "plot_bayesian_nomogram.py\n",
      "plot_bubble_map.py\n",
      "plot_color_text.py\n",
      "plot_context.py\n",
      "plot_distributions.py\n",
      "plot_heat_map.py\n",
      "plot_pie.py\n",
      "plot_points.py\n",
      "plot_table.py\n",
      "plot_violin_or_box.py\n",
      "process_feature_x_sample.py\n",
      "read_copynumber_gistic2.py\n",
      "read_correlate_copynumber_vs_mrnaseq.py\n",
      "read_gct.py\n",
      "read_gff3_gz.py\n",
      "read_gmt.py\n",
      "read_gmts.py\n",
      "read_json.py\n",
      "read_matrix_market.py\n",
      "read_mutsignozzlereport2cv.py\n",
      "read_vcf_gz_and_make_vcf_dict.py\n",
      "read_where_and_map_column_names.py\n",
      "reboot_machine.py\n",
      "remove_path.py\n",
      "remove_paths.py\n",
      "rename_chromosome_of_vcf_gz_using_bcftools_annotate.py\n",
      "replace_bad_objects_in_iterable.py\n",
      "rescale_x_y_coordiantes_in_polar_coordiante.py\n",
      "reverse_complement_dna_sequence.py\n",
      "reverse_transcribe_rna_sequence.py\n",
      "run_command.py\n",
      "run_command_and_monitor.py\n",
      "select_gene_symbol.py\n",
      "select_series_indices.py\n",
      "shuffle_each_2d_array_slice.py\n",
      "shutdown_machine.py\n",
      "simulate_sequences_using_dwgsim.py\n",
      "single_sample_gsea.py\n",
      "single_sample_gseas.py\n",
      "solve_ax_equal_b.py\n",
      "solve_for_H.py\n",
      "sort_and_index_bam_using_samtools_sort_and_index.py\n",
      "split_codons.py\n",
      "split_df.py\n",
      "split_maf_by_tumor_sample_barcode.py\n",
      "split_str_ignoring_inside_quotes.py\n",
      "str_is_version.py\n",
      "summarize_feature_x_sample.py\n",
      "title_str.py\n",
      "train_and_classify.py\n",
      "train_and_regress.py\n",
      "transcribe_dna_sequence.py\n",
      "translate_nucleotide_sequence.py\n",
      "trim_fastq_gzs_using_skewer.py\n",
      "unmount_volume.py\n",
      "untitle_str.py\n",
      "update_variant_dict.py\n",
      "write_dict.py\n",
      "write_gct.py\n",
      "write_gmt.py\n",
      "write_json.py\n"
     ]
    }
   ],
   "source": [
    "import_lines = []\n",
    "\n",
    "for py_file_name in sorted(os.listdir(\"../ccal\")):\n",
    "\n",
    "    print(py_file_name)\n",
    "\n",
    "    if py_file_name != \"__init__.py\" and py_file_name.endswith(\".py\"):\n",
    "\n",
    "        check_py_file_path(\"../ccal/{}\".format(py_file_name))\n",
    "\n",
    "        import_lines.append(\"from .{0} import {0}\".format(py_file_name[: -len(\".py\")]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../ccal/__init__.py\", mode=\"w\") as init_file:\n",
    "\n",
    "    init_file.write(\n",
    "        \"\\n\".join(\n",
    "            (\n",
    "                \"from os.path import abspath\",\n",
    "                'VERSION = \"{}\"'.format(version),\n",
    "                'print(\"CCAL version {} @ {}\".format(VERSION, abspath(__file__)))',\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "\n",
    "    init_file.write(\"\\n\")\n",
    "\n",
    "    init_file.writelines(\"\\n\".join(import_lines))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCAL version 0.9.1 @ /home/kwatme/github/ccal/ccal/__init__.py\n"
     ]
    }
   ],
   "source": [
    "sys.path.insert(0, \"..\")\n",
    "\n",
    "import ccal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "268\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['BAD_VARIANT_IDS',\n",
       " 'CODON_TO_AMINO_ACID',\n",
       " 'COLOR_CATEGORICAL',\n",
       " 'COLOR_RUBY_EMERALD',\n",
       " 'COLOR_WHITE_BLACK',\n",
       " 'COLOR_WHITE_BROWN',\n",
       " 'DATA_DIRECTORY_PATH',\n",
       " 'FeatureHDF5',\n",
       " 'GPSMap',\n",
       " 'Genome',\n",
       " 'VARIANT_CLASSIFICATION_MUTSIG_EFFECT',\n",
       " 'VARIANT_EFFECTS',\n",
       " 'VCF_ANN_FIELDS',\n",
       " 'VCF_COLUMNS',\n",
       " 'VERSION',\n",
       " 'VariantHDF5',\n",
       " '__builtins__',\n",
       " '__cached__',\n",
       " '__doc__',\n",
       " '__file__',\n",
       " '__loader__',\n",
       " '__name__',\n",
       " '__package__',\n",
       " '__path__',\n",
       " '__spec__',\n",
       " '_anneal_node_and_element_positions',\n",
       " '_check_fastq_gzs',\n",
       " '_check_node_x_element',\n",
       " '_check_w_or_h',\n",
       " '_cluster_clustering_x_element_and_compute_ccc',\n",
       " '_compute_context_indices',\n",
       " '_compute_norm',\n",
       " '_count',\n",
       " '_describe_vcf_df',\n",
       " '_fit_skew_t_pdfs',\n",
       " '_get_coclustering_portion',\n",
       " '_get_target_grid_indices',\n",
       " '_get_triangulation_edges',\n",
       " '_gzip_compress',\n",
       " '_identify_what_to_count',\n",
       " '_ignore_bad_and_compute_euclidean_distance_between_2_1d_arrays',\n",
       " '_make_annotations',\n",
       " '_make_clean_vcf_df',\n",
       " '_make_context_matrix',\n",
       " '_make_element_x_dimension',\n",
       " '_make_grid_values_and_categorical_labels',\n",
       " '_make_variant_dict_consistent',\n",
       " '_match',\n",
       " '_match_randomly_sampled_target_and_data_to_compute_margin_of_errors',\n",
       " '_match_target_and_data',\n",
       " '_normalize_nd_array',\n",
       " '_permute_target_and_match_target_and_data',\n",
       " '_plot_2d',\n",
       " '_plot_gps_map',\n",
       " '_plot_mountain',\n",
       " '_print_and_run_command',\n",
       " '_process_target_or_data_for_plotting',\n",
       " '_single_sample_gseas',\n",
       " '_update_H_by_multiplicative_update',\n",
       " '_update_W_by_multiplicative_update',\n",
       " 'abspath',\n",
       " 'add_conda_to_path',\n",
       " 'align_fastq_gzs_using_bwa_mem',\n",
       " 'align_fastq_gzs_using_hisat2',\n",
       " 'annotate_vcf_gz_using_bcftools_annotate',\n",
       " 'annotate_vcf_gz_using_snpeff',\n",
       " 'apply_function_on_2_1d_arrays',\n",
       " 'apply_function_on_2_2d_arrays_slices',\n",
       " 'bgzip_and_tabix',\n",
       " 'cast_series_to_builtins',\n",
       " 'cast_str_to_builtins',\n",
       " 'check_bam_using_samtools_flagstat',\n",
       " 'check_fastq_gzs_using_fastqc',\n",
       " 'check_nd_array_for_bad',\n",
       " 'clean_and_write_df_to_tsv',\n",
       " 'clean_git_url',\n",
       " 'clean_name',\n",
       " 'clean_path',\n",
       " 'clip_nd_array_by_standard_deviation',\n",
       " 'cluster_2d_array_slices',\n",
       " 'compute_bandwidths',\n",
       " 'compute_context',\n",
       " 'compute_correlation_distance',\n",
       " 'compute_empirical_p_value',\n",
       " 'compute_empirical_p_values_and_fdrs',\n",
       " 'compute_entropy',\n",
       " 'compute_information_coefficient',\n",
       " 'compute_information_distance',\n",
       " 'compute_joint_probability',\n",
       " 'compute_kullback_leibler_divergence',\n",
       " 'compute_mutational_signature_enrichment',\n",
       " 'compute_nd_array_margin_of_error',\n",
       " 'compute_posterior_probability',\n",
       " 'concatenate_vcf_gzs_using_bcftools_concat',\n",
       " 'conda_is_installed',\n",
       " 'copy_path',\n",
       " 'correlate',\n",
       " 'count_gene_impacts_from_variant_dicts',\n",
       " 'count_transcripts_using_kallisto_quant',\n",
       " 'count_vcf_gz_rows',\n",
       " 'create_gitkeep',\n",
       " 'cross_validate',\n",
       " 'download',\n",
       " 'download_and_parse_geo_data',\n",
       " 'download_clinvar_vcf_gz',\n",
       " 'drop_df_slice',\n",
       " 'drop_df_slice_greedily',\n",
       " 'dump_gps_map',\n",
       " 'echo_or_print',\n",
       " 'establish_fai_index',\n",
       " 'establish_path',\n",
       " 'estimate_kernel_density',\n",
       " 'exit_',\n",
       " 'faidx_fasta',\n",
       " 'filter_vcf_gz_using_bcftools_view',\n",
       " 'fit_skew_t_pdf',\n",
       " 'fit_skew_t_pdfs',\n",
       " 'flatten_nested_iterable',\n",
       " 'get_1d_array_unique_objects_in_order',\n",
       " 'get_allelic_frequencies',\n",
       " 'get_chromosome_size_from_fasta_gz',\n",
       " 'get_colormap_colors',\n",
       " 'get_conda_environments',\n",
       " 'get_conda_prefix',\n",
       " 'get_function_name',\n",
       " 'get_genotype',\n",
       " 'get_gff3_attribute',\n",
       " 'get_git_versions',\n",
       " 'get_installed_pip_libraries',\n",
       " 'get_intersections_between_2_1d_arrays',\n",
       " 'get_machine',\n",
       " 'get_maf_variant_classification',\n",
       " 'get_now',\n",
       " 'get_object_reference',\n",
       " 'get_open_port',\n",
       " 'get_population_allelic_frequencies',\n",
       " 'get_sequence_from_fasta_gz',\n",
       " 'get_shell_environment',\n",
       " 'get_unique_iterable_objects_in_order',\n",
       " 'get_variant_start_and_end_positions',\n",
       " 'get_variant_type',\n",
       " 'get_variants_from_bam_using_freebayes',\n",
       " 'get_variants_from_bam_using_freebayes_and_multiprocess',\n",
       " 'get_variants_from_bam_using_strelka',\n",
       " 'get_variants_from_vcf_gz',\n",
       " 'get_vcf_info',\n",
       " 'get_vcf_info_ann',\n",
       " 'get_vcf_sample_format',\n",
       " 'get_volume_name',\n",
       " 'group_and_apply_function_on_each_group_in_iterable',\n",
       " 'group_iterable',\n",
       " 'gsea',\n",
       " 'gzip_compress_file',\n",
       " 'gzip_decompress_and_bgzip_compress_file',\n",
       " 'gzip_decompress_file',\n",
       " 'have_program',\n",
       " 'hierarchical_consensus_cluster',\n",
       " 'hierarchical_consensus_cluster_with_ks',\n",
       " 'in_git_repository',\n",
       " 'index_bam_using_samtools_index',\n",
       " 'index_gff3_df_by_name',\n",
       " 'infer',\n",
       " 'infer_assuming_independence',\n",
       " 'initialize_logger',\n",
       " 'install_and_activate_conda',\n",
       " 'install_python_libraries',\n",
       " 'is_inframe',\n",
       " 'is_valid_vcf_gz',\n",
       " 'load_gps_map',\n",
       " 'log_and_return_response',\n",
       " 'log_nd_array',\n",
       " 'make_case_annotations',\n",
       " 'make_categorical_colors',\n",
       " 'make_colorscale',\n",
       " 'make_colorscale_from_colors',\n",
       " 'make_comparison_panel',\n",
       " 'make_context_matrix',\n",
       " 'make_coordinates_for_reflection',\n",
       " 'make_file_name_from_str',\n",
       " 'make_maf_from_vcf',\n",
       " 'make_match_panel',\n",
       " 'make_match_panels',\n",
       " 'make_membership_df_from_categorical_series',\n",
       " 'make_mesh_grid_coordinates_per_axis',\n",
       " 'make_object_int_mapping',\n",
       " 'make_random_color',\n",
       " 'make_reference_genome',\n",
       " 'make_summary_match_panel',\n",
       " 'make_volume_dict',\n",
       " 'map_cell_line_names',\n",
       " 'mark_duplicates_in_bam_using_picard_markduplicates',\n",
       " 'mds',\n",
       " 'merge_dicts_with_callable',\n",
       " 'mount_volume',\n",
       " 'multiprocess',\n",
       " 'nd_array_is_sorted',\n",
       " 'nmf_by_multiple_V_and_H',\n",
       " 'nmf_by_multiplicative_update',\n",
       " 'nmf_by_sklearn',\n",
       " 'nmf_consensus_cluster',\n",
       " 'nmf_consensus_cluster_with_ks',\n",
       " 'normalize_contig',\n",
       " 'normalize_df',\n",
       " 'normalize_nd_array',\n",
       " 'parse_vcf_row_and_make_variant_dict',\n",
       " 'plot_and_save',\n",
       " 'plot_bar',\n",
       " 'plot_bayesian_nomogram',\n",
       " 'plot_bubble_map',\n",
       " 'plot_color_text',\n",
       " 'plot_context',\n",
       " 'plot_distributions',\n",
       " 'plot_heat_map',\n",
       " 'plot_pie',\n",
       " 'plot_points',\n",
       " 'plot_table',\n",
       " 'plot_violin_or_box',\n",
       " 'process_feature_x_sample',\n",
       " 'read_copynumber_gistic2',\n",
       " 'read_correlate_copynumber_vs_mrnaseq',\n",
       " 'read_gct',\n",
       " 'read_gff3_gz',\n",
       " 'read_gmt',\n",
       " 'read_gmts',\n",
       " 'read_json',\n",
       " 'read_matrix_market',\n",
       " 'read_mutsignozzlereport2cv',\n",
       " 'read_vcf_gz_and_make_vcf_dict',\n",
       " 'read_where_and_map_column_names',\n",
       " 'reboot_machine',\n",
       " 'remove_path',\n",
       " 'remove_paths',\n",
       " 'rename_chromosome_of_vcf_gz_using_bcftools_annotate',\n",
       " 'replace_bad_objects_in_iterable',\n",
       " 'rescale_x_y_coordiantes_in_polar_coordiante',\n",
       " 'reverse_complement_dna_sequence',\n",
       " 'reverse_transcribe_rna_sequence',\n",
       " 'run_command',\n",
       " 'run_command_and_monitor',\n",
       " 'select_gene_symbol',\n",
       " 'select_series_indices',\n",
       " 'shuffle_each_2d_array_slice',\n",
       " 'shutdown_machine',\n",
       " 'simulate_sequences_using_dwgsim',\n",
       " 'single_sample_gsea',\n",
       " 'single_sample_gseas',\n",
       " 'solve_ax_equal_b',\n",
       " 'solve_for_H',\n",
       " 'sort_and_index_bam_using_samtools_sort_and_index',\n",
       " 'split_codons',\n",
       " 'split_df',\n",
       " 'split_maf_by_tumor_sample_barcode',\n",
       " 'split_str_ignoring_inside_quotes',\n",
       " 'str_is_version',\n",
       " 'summarize_feature_x_sample',\n",
       " 'title_str',\n",
       " 'train_and_classify',\n",
       " 'train_and_regress',\n",
       " 'transcribe_dna_sequence',\n",
       " 'translate_nucleotide_sequence',\n",
       " 'trim_fastq_gzs_using_skewer',\n",
       " 'unmount_volume',\n",
       " 'untitle_str',\n",
       " 'update_variant_dict',\n",
       " 'write_dict',\n",
       " 'write_gct',\n",
       " 'write_gmt',\n",
       " 'write_json']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(dir(ccal)))\n",
    "\n",
    "dir(ccal)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
