{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_importable_name_from_py_file_path(py_file_path):\n",
    "\n",
    "    py_file_name == py_file_path.split(\"/\")[-1]\n",
    "\n",
    "    py_file_name_without_py = py_file_name[: -len(\".py\")]\n",
    "\n",
    "    if py_file_name_without_py.isupper():\n",
    "\n",
    "        importable_names = [py_file_name_without_py]\n",
    "\n",
    "    else:\n",
    "\n",
    "        importable_names = []\n",
    "\n",
    "        with open(py_file_path) as py_file:\n",
    "\n",
    "            for line in py_file.readlines():\n",
    "\n",
    "                for prefix in (\"def \", \"class \"):\n",
    "\n",
    "                    if line.startswith(prefix):\n",
    "\n",
    "                        if \"(\" in line:\n",
    "\n",
    "                            separator = \"(\"\n",
    "\n",
    "                        elif \":\" in line:\n",
    "\n",
    "                            separator = \":\"\n",
    "\n",
    "                        importable_names.append(\n",
    "                            line.strip().split(separator)[0][len(prefix) :]\n",
    "                        )\n",
    "\n",
    "    _names = [name for name in importable_names if name.startswith(\"_\")]\n",
    "\n",
    "    if 0 < len(_names):\n",
    "\n",
    "        print(\"\\t\" + \"\\n\\t\".join(_names))\n",
    "\n",
    "    importable_names = set(importable_names) - set(_names)\n",
    "\n",
    "    if len(importable_names) != 1:\n",
    "\n",
    "        raise ValueError(\"N importable names ({}) != 1\".format(len(importable_names)))\n",
    "\n",
    "    importable_name = importable_names.pop()\n",
    "\n",
    "    if importable_name != py_file_name_without_py:\n",
    "\n",
    "        raise ValueError(\n",
    "            \"{}(.py) != {}\".format(py_file_name_without_py, importable_name)\n",
    "        )\n",
    "\n",
    "    return importable_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ALMOST_ZERO.py\n",
      "\n",
      "BAD_VALUES.py\n",
      "\n",
      "CODON_AMINO_ACID.py\n",
      "\n",
      "DATA_DIRECTORY_PATH.py\n",
      "\n",
      "FRACTION_GRID_EXTENSION.py\n",
      "\n",
      "FeatureAccessor.py\n",
      "\t_FeatureHDF5Description\n",
      "\n",
      "GPSMap.py\n",
      "\t_check_node_x_element\n",
      "\t_make_element_x_dimension\n",
      "\t_plot\n",
      "\n",
      "GSEA.py\n",
      "\n",
      "GSEAss.py\n",
      "\n",
      "GSEAss_and_match.py\n",
      "\n",
      "NONE_STRS.py\n",
      "\n",
      "N_GRID.py\n",
      "\n",
      "RANDOM_SEED.py\n",
      "\n",
      "VCF_ANN_KEYS.py\n",
      "\n",
      "VCF_COLUMNS.py\n",
      "\n",
      "VariantAccessor.py\n",
      "\t_VariantHDF5Description\n",
      "\n",
      "__init__.py\n",
      "\n",
      "__pycache__\n",
      "\n",
      "add_conda_to_path.py\n",
      "\n",
      "apply_function_on_2_1d_arrays.py\n",
      "\n",
      "apply_function_on_2_2d_arrays_slices.py\n",
      "\n",
      "apply_function_on_2_dataframes_slices.py\n",
      "\n",
      "call_function_with_multiprocess.py\n",
      "\n",
      "cast_object_to_builtin.py\n",
      "\n",
      "check_nd_array_for_bad.py\n",
      "\n",
      "clean_and_write_dataframe_to_tsv.py\n",
      "\n",
      "clip_nd_array_by_standard_deviation.py\n",
      "\n",
      "cluster_2d_array.py\n",
      "\t_compute_euclidean_distance\n",
      "\n",
      "cluster_clustering_x_element_and_compute_ccc.py\n",
      "\n",
      "compute_1d_array_bandwidth.py\n",
      "\n",
      "compute_1d_array_context.py\n",
      "\t_compute_pdf_context\n",
      "\n",
      "compute_1d_array_entropy.py\n",
      "\n",
      "compute_empirical_p_value.py\n",
      "\n",
      "compute_empirical_p_values_and_fdrs.py\n",
      "\n",
      "compute_information_coefficient_between_2_1d_arrays.py\n",
      "\n",
      "compute_information_distance_between_2_1d_arrays.py\n",
      "\n",
      "compute_joint_probability.py\n",
      "\n",
      "compute_kullback_leibler_divergence_between_2_pdfs.py\n",
      "\n",
      "compute_matrix_norm.py\n",
      "\n",
      "compute_normal_pdf_margin_of_error.py\n",
      "\n",
      "compute_posterior_probability.py\n",
      "\n",
      "correlate_2_1d_arrays.py\n",
      "\n",
      "count_gene_impacts_from_variant_dicts.py\n",
      "\n",
      "data\n",
      "\n",
      "download_and_parse_geo.py\n",
      "\n",
      "download_url.py\n",
      "\n",
      "download_url_and_extract.py\n",
      "\n",
      "drop_dataframe_slice.py\n",
      "\n",
      "drop_dataframe_slice_greedily.py\n",
      "\n",
      "echo_or_print_str.py\n",
      "\n",
      "establish_path.py\n",
      "\n",
      "estimate_kernel_density.py\n",
      "\n",
      "exit_.py\n",
      "\n",
      "fit_skew_t_pdf_on_1d_array.py\n",
      "\n",
      "fit_skew_t_pdf_on_each_dataframe_row.py\n",
      "\t_fit_skew_t_pdf_on_each_dataframe_row\n",
      "\n",
      "flatten_nested_iterable.py\n",
      "\n",
      "get_child_paths.py\n",
      "\n",
      "get_chromosome_size_from_fasta_gz.py\n",
      "\n",
      "get_colormap_colors.py\n",
      "\n",
      "get_conda_environments.py\n",
      "\n",
      "get_conda_prefix.py\n",
      "\n",
      "get_d_dimensions.py\n",
      "\n",
      "get_data_type.py\n",
      "\n",
      "get_gff3_attribute.py\n",
      "\n",
      "get_git_versions.py\n",
      "\n",
      "get_intersections_between_2_1d_arrays.py\n",
      "\n",
      "get_machine.py\n",
      "\n",
      "get_sequence_from_fasta_or_fasta_gz.py\n",
      "\n",
      "get_shell_environment.py\n",
      "\n",
      "get_triangulation_edges_from_point_x_dimension.py\n",
      "\n",
      "get_variant_start_and_end_positions.py\n",
      "\n",
      "get_variants_from_vcf_or_vcf_gz.py\n",
      "\n",
      "get_vcf_genotype.py\n",
      "\n",
      "get_vcf_info.py\n",
      "\n",
      "get_vcf_info_ann.py\n",
      "\n",
      "get_vcf_sample_format.py\n",
      "\n",
      "group_and_apply_function_on_each_group_in_iterable.py\n",
      "\n",
      "hierarchical_consensus_cluster.py\n",
      "\n",
      "hierarchical_consensus_cluster_with_ks.py\n",
      "\n",
      "infer.py\n",
      "\n",
      "infer_assuming_independence.py\n",
      "\n",
      "initialize_logger.py\n",
      "\n",
      "install_and_activate_conda.py\n",
      "\n",
      "install_python_libraries.py\n",
      "\n",
      "is_conda_directory_path.py\n",
      "\n",
      "is_in_git_repository.py\n",
      "\n",
      "is_program.py\n",
      "\n",
      "is_sorted_nd_array.py\n",
      "\n",
      "is_str_version.py\n",
      "\n",
      "isolate_information_x_sample.py\n",
      "\n",
      "log_and_return_response.py\n",
      "\n",
      "log_nd_array.py\n",
      "\n",
      "make_1d_array_grid.py\n",
      "\n",
      "make_binary_dataframe_from_categorical_series.py\n",
      "\n",
      "make_colorscale_from_colors.py\n",
      "\n",
      "make_context_matrix.py\n",
      "\t_make_context_matrix\n",
      "\n",
      "make_gitkeep.py\n",
      "\n",
      "make_match_panel.py\n",
      "\t_match_target_and_data\n",
      "\t_permute_target_and_match_target_and_data\n",
      "\t_match_target_and_data_and_compute_statistics\n",
      "\n",
      "make_match_panel_annotations.py\n",
      "\n",
      "make_match_panels.py\n",
      "\n",
      "make_mesh_grid_point_x_dimension.py\n",
      "\n",
      "make_reflecting_grid.py\n",
      "\n",
      "make_summary_match_panel.py\n",
      "\n",
      "make_variant_dict_consistent.py\n",
      "\n",
      "make_variant_dict_from_vcf_row.py\n",
      "\n",
      "make_variant_n_from_vcf_file_path.py\n",
      "\n",
      "make_variant_n_from_vcf_row.py\n",
      "\n",
      "map_objects_to_ints.py\n",
      "\n",
      "merge_2_dicts_with_function.py\n",
      "\n",
      "mf_by_multiple_V_and_H.py\n",
      "\n",
      "mf_by_multiplicative_update.py\n",
      "\n",
      "mf_consensus_cluster.py\n",
      "\n",
      "mf_consensus_cluster_with_ks.py\n",
      "\n",
      "nmf_by_sklearn.py\n",
      "\n",
      "normalize_cell_line_names.py\n",
      "\n",
      "normalize_contig.py\n",
      "\n",
      "normalize_file_name.py\n",
      "\n",
      "normalize_git_url.py\n",
      "\n",
      "normalize_nd_array.py\n",
      "\t_normalize_nd_array\n",
      "\n",
      "normalize_path.py\n",
      "\n",
      "normalize_series_or_dataframe.py\n",
      "\n",
      "pick_colors.py\n",
      "\n",
      "plot_bayesian_nomogram.py\n",
      "\n",
      "plot_bubble_map.py\n",
      "\n",
      "plot_context.py\n",
      "\n",
      "plot_heat_map.py\n",
      "\n",
      "plot_histogram.py\n",
      "\n",
      "plot_mesh_grid.py\n",
      "\n",
      "plot_plotly_figure.py\n",
      "\n",
      "plot_scatter_and_annotate.py\n",
      "\n",
      "print_function_information.py\n",
      "\t_function\n",
      "\n",
      "print_header_in_terminal.py\n",
      "\n",
      "process_feature_x_sample.py\n",
      "\n",
      "read_firebrowse_copynumber_gistic2.py\n",
      "\n",
      "read_firebrowse_correlate_copynumber_vs_mrnaseq.py\n",
      "\n",
      "read_firebrowse_mutsignozzlereport2cv.py\n",
      "\n",
      "read_gff3.py\n",
      "\n",
      "read_gmt.py\n",
      "\n",
      "read_gps_map.py\n",
      "\n",
      "read_json.py\n",
      "\n",
      "read_matrix_market.py\n",
      "\n",
      "read_where_and_map_column_name_on_hdf5_table.py\n",
      "\n",
      "remove_path.py\n",
      "\n",
      "rescale_x_y_coordiantes_in_polar_coordiante.py\n",
      "\n",
      "reverse_complement_dna_sequence.py\n",
      "\n",
      "reverse_transcribe_rna_sequence.py\n",
      "\n",
      "run_command.py\n",
      "\n",
      "run_gsea.py\n",
      "\n",
      "run_single_sample_gsea.py\n",
      "\n",
      "sample_from_each_series_value.py\n",
      "\n",
      "scale_point_x_dimension_dimension.py\n",
      "\n",
      "select_and_group_feature_x_tcga_sample_by_sample_type.py\n",
      "\n",
      "select_gene_symbol.py\n",
      "\n",
      "select_series_indices.py\n",
      "\n",
      "shuffle_each_2d_array_slice.py\n",
      "\n",
      "skip_quote_and_split_str.py\n",
      "\n",
      "solve_ax_equal_b.py\n",
      "\n",
      "solve_for_H.py\n",
      "\n",
      "split_codons.py\n",
      "\n",
      "split_dataframe.py\n",
      "\n",
      "summarize_feature_x_sample.py\n",
      "\n",
      "title_str.py\n",
      "\n",
      "train_and_classify.py\n",
      "\n",
      "train_and_regress.py\n",
      "\n",
      "transcribe_dna_sequence.py\n",
      "\n",
      "translate_nucleotide_sequence.py\n",
      "\n",
      "unmesh.py\n",
      "\n",
      "untitle_str.py\n",
      "\n",
      "update_H_by_multiplicative_update.py\n",
      "\n",
      "update_W_by_multiplicative_update.py\n",
      "\n",
      "update_variant_dict.py\n",
      "\n",
      "write_gps_map.py\n",
      "\n",
      "write_json.py\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import_lines = []\n",
    "\n",
    "for py_file_name in sorted(os.listdir(path=os.path.join(\"..\", \"kraft\"))):\n",
    "\n",
    "    print(py_file_name)\n",
    "\n",
    "    if py_file_name != \"__init__.py\" and py_file_name.endswith(\".py\"):\n",
    "\n",
    "        import_lines.append(\n",
    "            \"from .{0} import {0}\".format(\n",
    "                get_importable_name_from_py_file_path(\n",
    "                    os.path.join(\"..\", \"kraft\", py_file_name)\n",
    "                )\n",
    "            )\n",
    "        )\n",
    "\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(\"..\", \"kraft\", \"__init__.py\"), mode=\"w\") as io:\n",
    "\n",
    "    io.write('from plotly.io import templates\\ntemplates.default = \"plotly_white\"\\n')\n",
    "\n",
    "    io.writelines(\"\\n\".join(import_lines))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import kraft"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "180\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['ALMOST_ZERO',\n",
       " 'BAD_VALUES',\n",
       " 'CODON_AMINO_ACID',\n",
       " 'DATA_DIRECTORY_PATH',\n",
       " 'FRACTION_GRID_EXTENSION',\n",
       " 'FeatureAccessor',\n",
       " 'GPSMap',\n",
       " 'GSEA',\n",
       " 'GSEAss',\n",
       " 'GSEAss_and_match',\n",
       " 'NONE_STRS',\n",
       " 'N_GRID',\n",
       " 'RANDOM_SEED',\n",
       " 'VCF_ANN_KEYS',\n",
       " 'VCF_COLUMNS',\n",
       " 'VariantAccessor',\n",
       " '__builtins__',\n",
       " '__cached__',\n",
       " '__doc__',\n",
       " '__file__',\n",
       " '__loader__',\n",
       " '__name__',\n",
       " '__package__',\n",
       " '__path__',\n",
       " '__spec__',\n",
       " 'add_conda_to_path',\n",
       " 'apply_function_on_2_1d_arrays',\n",
       " 'apply_function_on_2_2d_arrays_slices',\n",
       " 'apply_function_on_2_dataframes_slices',\n",
       " 'call_function_with_multiprocess',\n",
       " 'cast_object_to_builtin',\n",
       " 'check_nd_array_for_bad',\n",
       " 'clean_and_write_dataframe_to_tsv',\n",
       " 'clip_nd_array_by_standard_deviation',\n",
       " 'cluster_2d_array',\n",
       " 'cluster_clustering_x_element_and_compute_ccc',\n",
       " 'compute_1d_array_bandwidth',\n",
       " 'compute_1d_array_context',\n",
       " 'compute_1d_array_entropy',\n",
       " 'compute_empirical_p_value',\n",
       " 'compute_empirical_p_values_and_fdrs',\n",
       " 'compute_information_coefficient_between_2_1d_arrays',\n",
       " 'compute_information_distance_between_2_1d_arrays',\n",
       " 'compute_joint_probability',\n",
       " 'compute_kullback_leibler_divergence_between_2_pdfs',\n",
       " 'compute_matrix_norm',\n",
       " 'compute_normal_pdf_margin_of_error',\n",
       " 'compute_posterior_probability',\n",
       " 'correlate_2_1d_arrays',\n",
       " 'count_gene_impacts_from_variant_dicts',\n",
       " 'download_and_parse_geo',\n",
       " 'download_url',\n",
       " 'download_url_and_extract',\n",
       " 'drop_dataframe_slice',\n",
       " 'drop_dataframe_slice_greedily',\n",
       " 'echo_or_print_str',\n",
       " 'establish_path',\n",
       " 'estimate_kernel_density',\n",
       " 'exit_',\n",
       " 'fit_skew_t_pdf_on_1d_array',\n",
       " 'fit_skew_t_pdf_on_each_dataframe_row',\n",
       " 'flatten_nested_iterable',\n",
       " 'get_child_paths',\n",
       " 'get_chromosome_size_from_fasta_gz',\n",
       " 'get_colormap_colors',\n",
       " 'get_conda_environments',\n",
       " 'get_conda_prefix',\n",
       " 'get_d_dimensions',\n",
       " 'get_data_type',\n",
       " 'get_gff3_attribute',\n",
       " 'get_git_versions',\n",
       " 'get_intersections_between_2_1d_arrays',\n",
       " 'get_machine',\n",
       " 'get_sequence_from_fasta_or_fasta_gz',\n",
       " 'get_shell_environment',\n",
       " 'get_triangulation_edges_from_point_x_dimension',\n",
       " 'get_variant_start_and_end_positions',\n",
       " 'get_variants_from_vcf_or_vcf_gz',\n",
       " 'get_vcf_genotype',\n",
       " 'get_vcf_info',\n",
       " 'get_vcf_info_ann',\n",
       " 'get_vcf_sample_format',\n",
       " 'group_and_apply_function_on_each_group_in_iterable',\n",
       " 'hierarchical_consensus_cluster',\n",
       " 'hierarchical_consensus_cluster_with_ks',\n",
       " 'infer',\n",
       " 'infer_assuming_independence',\n",
       " 'initialize_logger',\n",
       " 'install_and_activate_conda',\n",
       " 'install_python_libraries',\n",
       " 'is_conda_directory_path',\n",
       " 'is_in_git_repository',\n",
       " 'is_program',\n",
       " 'is_sorted_nd_array',\n",
       " 'is_str_version',\n",
       " 'isolate_information_x_sample',\n",
       " 'log_and_return_response',\n",
       " 'log_nd_array',\n",
       " 'make_1d_array_grid',\n",
       " 'make_binary_dataframe_from_categorical_series',\n",
       " 'make_colorscale_from_colors',\n",
       " 'make_context_matrix',\n",
       " 'make_gitkeep',\n",
       " 'make_match_panel',\n",
       " 'make_match_panel_annotations',\n",
       " 'make_match_panels',\n",
       " 'make_mesh_grid_point_x_dimension',\n",
       " 'make_reflecting_grid',\n",
       " 'make_summary_match_panel',\n",
       " 'make_variant_dict_consistent',\n",
       " 'make_variant_dict_from_vcf_row',\n",
       " 'make_variant_n_from_vcf_file_path',\n",
       " 'make_variant_n_from_vcf_row',\n",
       " 'map_objects_to_ints',\n",
       " 'merge_2_dicts_with_function',\n",
       " 'mf_by_multiple_V_and_H',\n",
       " 'mf_by_multiplicative_update',\n",
       " 'mf_consensus_cluster',\n",
       " 'mf_consensus_cluster_with_ks',\n",
       " 'nmf_by_sklearn',\n",
       " 'normalize_cell_line_names',\n",
       " 'normalize_contig',\n",
       " 'normalize_file_name',\n",
       " 'normalize_git_url',\n",
       " 'normalize_nd_array',\n",
       " 'normalize_path',\n",
       " 'normalize_series_or_dataframe',\n",
       " 'pick_colors',\n",
       " 'plot_bayesian_nomogram',\n",
       " 'plot_bubble_map',\n",
       " 'plot_context',\n",
       " 'plot_heat_map',\n",
       " 'plot_histogram',\n",
       " 'plot_mesh_grid',\n",
       " 'plot_plotly_figure',\n",
       " 'plot_scatter_and_annotate',\n",
       " 'print_function_information',\n",
       " 'print_header_in_terminal',\n",
       " 'process_feature_x_sample',\n",
       " 'read_firebrowse_copynumber_gistic2',\n",
       " 'read_firebrowse_correlate_copynumber_vs_mrnaseq',\n",
       " 'read_firebrowse_mutsignozzlereport2cv',\n",
       " 'read_gff3',\n",
       " 'read_gmt',\n",
       " 'read_gps_map',\n",
       " 'read_json',\n",
       " 'read_matrix_market',\n",
       " 'read_where_and_map_column_name_on_hdf5_table',\n",
       " 'remove_path',\n",
       " 'rescale_x_y_coordiantes_in_polar_coordiante',\n",
       " 'reverse_complement_dna_sequence',\n",
       " 'reverse_transcribe_rna_sequence',\n",
       " 'run_command',\n",
       " 'run_gsea',\n",
       " 'run_single_sample_gsea',\n",
       " 'sample_from_each_series_value',\n",
       " 'scale_point_x_dimension_dimension',\n",
       " 'select_and_group_feature_x_tcga_sample_by_sample_type',\n",
       " 'select_gene_symbol',\n",
       " 'select_series_indices',\n",
       " 'shuffle_each_2d_array_slice',\n",
       " 'skip_quote_and_split_str',\n",
       " 'solve_ax_equal_b',\n",
       " 'solve_for_H',\n",
       " 'split_codons',\n",
       " 'split_dataframe',\n",
       " 'summarize_feature_x_sample',\n",
       " 'templates',\n",
       " 'title_str',\n",
       " 'train_and_classify',\n",
       " 'train_and_regress',\n",
       " 'transcribe_dna_sequence',\n",
       " 'translate_nucleotide_sequence',\n",
       " 'unmesh',\n",
       " 'untitle_str',\n",
       " 'update_H_by_multiplicative_update',\n",
       " 'update_W_by_multiplicative_update',\n",
       " 'update_variant_dict',\n",
       " 'write_gps_map',\n",
       " 'write_json']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(dir(kraft)))\n",
    "\n",
    "dir(kraft)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, file_path in enumerate(sorted(os.listdir())):\n",
    "\n",
    "    if file_path.endswith(\".ipynb\") and not file_path.startswith(\"_\"):\n",
    "\n",
    "        print(i, file_path)\n",
    "\n",
    "        subprocess.run(\n",
    "            \"jupyter nbconvert --execute --ExecutePreprocessor.timeout=-1 --inplace {}\".format(\n",
    "                file_path\n",
    "            ),\n",
    "            shell=True,\n",
    "            check=True,\n",
    "        )\n",
    "\n",
    "        subprocess.run(\"clean_ipynb {}\".format(file_path), shell=True, check=True)"
   ]
  }
 ],
 "metadata": {
  "@webio": {
   "lastCommId": null,
   "lastKernelId": null
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
