{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-06T03:08:08.440048Z",
     "start_time": "2018-05-06T03:08:08.435655Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-06T03:08:08.447578Z",
     "start_time": "2018-05-06T03:08:08.443639Z"
    }
   },
   "outputs": [],
   "source": [
    "version = '0.6.7'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-06T03:08:08.547901Z",
     "start_time": "2018-05-06T03:08:08.455373Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipped ../ccal/nmf_map/notebook.\n",
      "Skipped ../ccal/cluster/notebook.\n",
      "Skipped ../ccal/information/notebook.\n",
      "Skipped ../ccal/context/notebook.\n",
      "Skipped ../ccal/classify/notebook.\n",
      "Skipped ../ccal/probability/notebook.\n",
      "Skipped ../ccal/plot/notebook.\n",
      "Skipped ../ccal/gsea/notebook.\n",
      "Skipped ../ccal/regress/notebook.\n",
      "Skipped ../ccal/match/notebook.\n",
      "Skipped ../ccal/matrix_factorization/notebook.\n",
      "Skipped ../ccal/geo/notebook.\n",
      "Skipped ../ccal/gct_gmt/notebook.\n",
      "Skipped ../ccal/linear_model/notebook.\n",
      "Skipped ../ccal/mutational_signature/media.\n",
      "Skipped ../ccal/cryptograph/notebook.\n",
      "Skipped ../ccal/support/notebook.\n",
      "Skipped ../ccal/dimension_scale/notebook.\n",
      "Skipped ../ccal/variant/notebook.\n",
      "Skipped ../ccal/kernel_density/notebook.\n",
      "Skipped ../ccal/cross_validate/notebook.\n",
      "Skipped ../ccal/feature/notebook.\n",
      "Skipped ../ccal/sequencing_process/notebook.\n",
      "Skipped ../ccal/sequencing_process/resource.\n",
      "Skipped ../ccal/genome/notebook.\n",
      "Skipped ../ccal/nd_array/notebook.\n"
     ]
    }
   ],
   "source": [
    "imports = []\n",
    "\n",
    "for dp, dns, fns in os.walk('../ccal'):\n",
    "\n",
    "    split = dp.split('/')\n",
    "\n",
    "    if len(split) == 4:\n",
    "\n",
    "        if split[2] == split[3]:\n",
    "\n",
    "            for fn in fns:\n",
    "\n",
    "                if fn.endswith('.py'):\n",
    "\n",
    "                    with open(os.path.join(dp, fn)) as f:\n",
    "\n",
    "                        for l in f.readlines():\n",
    "\n",
    "                            if l.startswith('def '):\n",
    "\n",
    "                                l = l.strip()\n",
    "\n",
    "                                function_name = l.split('(')[0][4:]\n",
    "\n",
    "                                if not function_name.startswith('_'):\n",
    "\n",
    "                                    imports.append(\n",
    "                                        'from .{0}.{0}.{1} import {2}'.format(\n",
    "                                            split[2], fn[:-3], function_name))\n",
    "\n",
    "                            elif l.startswith('class '):\n",
    "\n",
    "                                l = l.strip()\n",
    "\n",
    "                                class_name = l.split('(')[0][6:-1]\n",
    "\n",
    "                                imports.append(\n",
    "                                    'from .{0}.{0}.{1} import {2}'.format(\n",
    "                                        split[2], fn[:-3], class_name))\n",
    "\n",
    "        else:\n",
    "\n",
    "            print('Skipped {}.'.format(os.path.join(*split)))\n",
    "\n",
    "spacer = '\\n\\n'\n",
    "\n",
    "with open('../ccal/__init__.py', 'w') as file:\n",
    "\n",
    "    file.write('from os.path import abspath')\n",
    "\n",
    "    file.write(spacer)\n",
    "\n",
    "    file.write('VERSION = \\'{}\\''.format(version))\n",
    "\n",
    "    file.write(spacer)\n",
    "\n",
    "    file.write(\n",
    "        'print(\\'CCAL version {} @ {}\\'.format(VERSION, abspath(__file__)))')\n",
    "\n",
    "    file.write(spacer)\n",
    "\n",
    "    file.write('\\n'.join(imports))\n",
    "\n",
    "    file.write(spacer)\n",
    "    file.write('\\n')\n",
    "\n",
    "    file.write('explore_components = nmf_consensus_cluster_with_multiple_k')\n",
    "\n",
    "    file.write(spacer)\n",
    "\n",
    "    file.write(\n",
    "        'explore_states = hierarchical_consensus_cluster_with_multiple_k')\n",
    "\n",
    "    file.write(spacer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-06T03:08:11.605505Z",
     "start_time": "2018-05-06T03:08:08.551539Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCAL version 0.6.7 @ /Users/k/Jumis/github_ccal/ccal/ccal/__init__.py\n"
     ]
    }
   ],
   "source": [
    "sys.path.insert(0, '..')\n",
    "\n",
    "import ccal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-06T03:08:11.647883Z",
     "start_time": "2018-05-06T03:08:11.610623Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['FeatureHDF5',\n",
       " 'Genome',\n",
       " 'NMFMap',\n",
       " 'VERSION',\n",
       " 'VariantHDF5',\n",
       " '__builtins__',\n",
       " '__cached__',\n",
       " '__doc__',\n",
       " '__file__',\n",
       " '__loader__',\n",
       " '__name__',\n",
       " '__package__',\n",
       " '__path__',\n",
       " '__spec__',\n",
       " 'abspath',\n",
       " 'add_conda_to_path',\n",
       " 'align_fastq_gzs_using_bwa_mem',\n",
       " 'align_fastq_gzs_using_hisat2',\n",
       " 'anneal_node_and_element_positions',\n",
       " 'annotate_vcf_gz_using_bcftools_annotate',\n",
       " 'annotate_vcf_gz_using_snpeff',\n",
       " 'apply_function_on_2_1d_arrays_and_compute_empirical_p_value',\n",
       " 'apply_function_on_2_2d_arrays_slices',\n",
       " 'bgzip_and_tabix',\n",
       " 'cast_series_to_builtins',\n",
       " 'cast_str_to_builtins',\n",
       " 'check_bam_using_samtools_flagstat',\n",
       " 'check_fastq_gz_or_bam_using_fastqp',\n",
       " 'check_fastq_gzs',\n",
       " 'check_fastq_gzs_using_fastqc',\n",
       " 'check_nd_array_for_bad_value',\n",
       " 'classify',\n",
       " 'clean_git_url',\n",
       " 'clean_name',\n",
       " 'clean_path',\n",
       " 'clip_nd_array_by_standard_deviation',\n",
       " 'cluster',\n",
       " 'cluster_1d_arrays',\n",
       " 'cluster_2d_array_slices_by_group',\n",
       " 'compute_bandwidths',\n",
       " 'compute_context',\n",
       " 'compute_cosine_similarity',\n",
       " 'compute_empirical_p_value',\n",
       " 'compute_empirical_p_values_and_fdrs',\n",
       " 'compute_entropy',\n",
       " 'compute_euclidean_distance',\n",
       " 'compute_gene_scores',\n",
       " 'compute_information_coefficient',\n",
       " 'compute_information_distance',\n",
       " 'compute_joint_probability',\n",
       " 'compute_log_ratio_between_2_1d_arrays',\n",
       " 'compute_mutational_signature_enrichment',\n",
       " 'compute_nd_array_margin_of_error',\n",
       " 'compute_pearsons_correlation',\n",
       " 'compute_posterior_probability',\n",
       " 'concatenate_vcf_gzs_using_bcftools_concat',\n",
       " 'conda_is_installed',\n",
       " 'context',\n",
       " 'copy_path',\n",
       " 'correlate',\n",
       " 'count_coclustering_and_normalize',\n",
       " 'count_gene_impacts_from_variant_dicts',\n",
       " 'count_transcripts_using_kallisto_quant',\n",
       " 'count_vcf_gz_rows',\n",
       " 'create_gitkeep',\n",
       " 'cross_validate',\n",
       " 'cryptograph',\n",
       " 'decrypt_directory_files',\n",
       " 'decrypt_file',\n",
       " 'dimension_scale',\n",
       " 'download',\n",
       " 'download_clinvar_vcf_gz',\n",
       " 'drop_bad_value_and_apply_function_on_2_1d_arrays',\n",
       " 'drop_df_na_greedily',\n",
       " 'drop_df_slice',\n",
       " 'drop_df_slice_greedily',\n",
       " 'echo_or_print',\n",
       " 'encrypt_directory_files',\n",
       " 'encrypt_file',\n",
       " 'establish_fai_index',\n",
       " 'establish_path',\n",
       " 'estimate_kernel_density',\n",
       " 'exit_',\n",
       " 'explore_components',\n",
       " 'explore_states',\n",
       " 'feature',\n",
       " 'filter_vcf_gz_using_bcftools_view',\n",
       " 'fit_skew_t_pdf',\n",
       " 'fit_skew_t_pdfs',\n",
       " 'flatten_nested_iterable',\n",
       " 'gct_gmt',\n",
       " 'genome',\n",
       " 'geo',\n",
       " 'get_1d_array_unique_objects_in_order',\n",
       " 'get_allelic_frequencies',\n",
       " 'get_and_parse_geo_data',\n",
       " 'get_chromosome_size_from_fasta_gz',\n",
       " 'get_conda_environments',\n",
       " 'get_conda_prefix',\n",
       " 'get_coordinates_for_reflection',\n",
       " 'get_elements_near_state_centroids',\n",
       " 'get_extreme_series_indices',\n",
       " 'get_genotype',\n",
       " 'get_gff3_attribute',\n",
       " 'get_git_versions',\n",
       " 'get_installed_pip_libraries',\n",
       " 'get_intersections_between_2_1d_arrays',\n",
       " 'get_machine',\n",
       " 'get_maf_variant_classification',\n",
       " 'get_now',\n",
       " 'get_object_reference',\n",
       " 'get_open_port',\n",
       " 'get_population_allelic_frequencies',\n",
       " 'get_sequence_from_fasta_gz',\n",
       " 'get_shell_environment',\n",
       " 'get_target_grid_indices',\n",
       " 'get_triangulation_edges',\n",
       " 'get_unique_iterable_objects_in_order',\n",
       " 'get_variant_start_and_end_positions',\n",
       " 'get_variant_type',\n",
       " 'get_variants_from_bam_using_freebayes',\n",
       " 'get_variants_from_bam_using_freebayes_and_multiprocess',\n",
       " 'get_variants_from_bam_using_strelka',\n",
       " 'get_variants_from_vcf_gz',\n",
       " 'get_vcf_info',\n",
       " 'get_vcf_info_ann',\n",
       " 'get_vcf_sample_format',\n",
       " 'get_volume_name',\n",
       " 'group_and_apply_function_on_each_group_in_iterable',\n",
       " 'group_iterable',\n",
       " 'gsea',\n",
       " 'gzip_compress_file',\n",
       " 'gzip_decompress_and_bgzip_compress_file',\n",
       " 'gzip_decompress_file',\n",
       " 'hash_directory_files',\n",
       " 'hash_file',\n",
       " 'hash_list_of_str',\n",
       " 'hash_passcode',\n",
       " 'hash_str_to_32_bytes',\n",
       " 'have_program',\n",
       " 'hdf5',\n",
       " 'hierarchical_consensus_cluster',\n",
       " 'hierarchical_consensus_cluster_with_multiple_k',\n",
       " 'in_git_repository',\n",
       " 'index_bam_using_samtools_index',\n",
       " 'index_gff3_df_by_name',\n",
       " 'infer',\n",
       " 'infer_assuming_independence',\n",
       " 'information',\n",
       " 'initialize_logger',\n",
       " 'install_and_activate_conda',\n",
       " 'install_python_libraries',\n",
       " 'is_inframe',\n",
       " 'is_valid_vcf_gz',\n",
       " 'kernel_density',\n",
       " 'linear_algebra',\n",
       " 'linear_model',\n",
       " 'log_and_return_response',\n",
       " 'make_1d_context_matrix',\n",
       " 'make_2d_context_matrix',\n",
       " 'make_annotations',\n",
       " 'make_case_annotations',\n",
       " 'make_colorscale',\n",
       " 'make_comparison_panel',\n",
       " 'make_maf_from_vcf',\n",
       " 'make_match_panel',\n",
       " 'make_membership_df_from_categorical_series',\n",
       " 'make_nd_grid_coordinates',\n",
       " 'make_object_int_mapping',\n",
       " 'make_random_color',\n",
       " 'make_reference_genome',\n",
       " 'make_summary_match_panel',\n",
       " 'make_volume_dict',\n",
       " 'makedictkey',\n",
       " 'makereversecomplementdictkey',\n",
       " 'makesigdict',\n",
       " 'mark_duplicates_in_bam_using_picard_markduplicates',\n",
       " 'match',\n",
       " 'match_passcode_with_passcode_hash',\n",
       " 'match_randomly_sampled_target_and_features_to_compute_margin_of_errors',\n",
       " 'match_target_and_features',\n",
       " 'matrix_factorization',\n",
       " 'mds',\n",
       " 'merge_dicts_with_callable',\n",
       " 'mount_volume',\n",
       " 'multiprocess',\n",
       " 'mutational_signature',\n",
       " 'nd_array',\n",
       " 'nd_array_is_sorted',\n",
       " 'nmf',\n",
       " 'nmf_consensus_cluster',\n",
       " 'nmf_consensus_cluster_with_multiple_k',\n",
       " 'nmf_map',\n",
       " 'normalize_nd_array',\n",
       " 'parse_vcf_row_and_make_variant_dict',\n",
       " 'permute_target_and_match_target_and_features',\n",
       " 'plot',\n",
       " 'plot_2d',\n",
       " 'plot_and_save',\n",
       " 'plot_bar',\n",
       " 'plot_bayesian_nomogram',\n",
       " 'plot_box_or_violin',\n",
       " 'plot_bubble_map',\n",
       " 'plot_context',\n",
       " 'plot_distributions',\n",
       " 'plot_heat_map',\n",
       " 'plot_mountain_plot',\n",
       " 'plot_nmf_map',\n",
       " 'plot_pie',\n",
       " 'plot_points',\n",
       " 'plot_table',\n",
       " 'print_and_run_command',\n",
       " 'probability',\n",
       " 'process_target_or_features_for_plotting',\n",
       " 'read_copynumber_gistic2',\n",
       " 'read_correlate_copynumber_vs_mrnaseq',\n",
       " 'read_gct',\n",
       " 'read_gff3',\n",
       " 'read_gmts',\n",
       " 'read_json',\n",
       " 'read_mutsignozzlereport2cv',\n",
       " 'read_vcf_gz_and_make_vcf_dict',\n",
       " 'read_where_and_map_column_names',\n",
       " 'reboot_machine',\n",
       " 'regress',\n",
       " 'remove_path',\n",
       " 'remove_paths',\n",
       " 'rename_chromosome_of_vcf_gz_using_bcftools_annotate',\n",
       " 'replace_bad_objects_in_iterable',\n",
       " 'rescale_x_y_coordiantes_in_polar_coordiante',\n",
       " 'reverse_complement_dna_sequence',\n",
       " 'reverse_transcribe_rna_sequence',\n",
       " 'reversecomplement',\n",
       " 'run_command',\n",
       " 'run_command_and_monitor',\n",
       " 'sequence',\n",
       " 'sequencing_process',\n",
       " 'shift_and_log_nd_array',\n",
       " 'shuffle_each_2d_array_slice',\n",
       " 'shutdown_machine',\n",
       " 'simulate_sequences_using_dwgsim',\n",
       " 'single_sample_gsea',\n",
       " 'single_sample_gseas',\n",
       " 'solve_ax_equal_b',\n",
       " 'solve_for_nmf_h',\n",
       " 'sort_and_index_bam_using_samtools_sort_and_index',\n",
       " 'split_codons',\n",
       " 'split_df',\n",
       " 'split_maf_by_tumor_sample_barcode',\n",
       " 'split_str_ignoring_inside_quotes',\n",
       " 'str_is_version',\n",
       " 'support',\n",
       " 'tcga',\n",
       " 'title_str',\n",
       " 'train_and_classify',\n",
       " 'train_and_regress',\n",
       " 'transcribe_dna_sequence',\n",
       " 'translate_nucleotide_sequence',\n",
       " 'trim_fastq_gzs_using_skewer',\n",
       " 'unmount_volume',\n",
       " 'untitle_str',\n",
       " 'update_variant_dict',\n",
       " 'variant',\n",
       " 'write_dict',\n",
       " 'write_gct',\n",
       " 'write_gmt',\n",
       " 'write_json']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(ccal)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
