{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-05-15T09:28:54.601864Z",
     "start_time": "2018-05-15T09:28:54.554930Z"
    }
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "version = \"0.9.0\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_importable_names_from_py_file_path(py_file_path):\n",
    "\n",
    "    importable_names = []\n",
    "\n",
    "    with open(py_file_path) as py_file:\n",
    "\n",
    "        for line in py_file.readlines():\n",
    "\n",
    "            for prefix, separator in ((\"def \", \"(\"), (\"class \", \":\")):\n",
    "\n",
    "                if line.startswith(prefix):\n",
    "\n",
    "                    importable_name = line.strip().split(separator)[0][len(prefix) :]\n",
    "\n",
    "                    if not importable_name.startswith(\"_\"):\n",
    "\n",
    "                        importable_names.append(importable_name)\n",
    "\n",
    "    return importable_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "================================================================================\n",
      "plot_bar.py\n",
      "\tplot_bar\n",
      "\n",
      "================================================================================\n",
      "dict_.py\n",
      "\tmerge_dicts_with_callable\n",
      "\n",
      "================================================================================\n",
      "estimate_kernel_density.py\n",
      "\testimate_kernel_density\n",
      "\n",
      "================================================================================\n",
      "_make_grid_values_and_categorical_labels.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "make_colorscale_from_colors.py\n",
      "\tmake_colorscale_from_colors\n",
      "\n",
      "================================================================================\n",
      "compression.py\n",
      "\tgzip_compress_file\n",
      "\tgzip_decompress_file\n",
      "\tgzip_decompress_and_bgzip_compress_file\n",
      "\n",
      "================================================================================\n",
      "compute_context.py\n",
      "\tcompute_context\n",
      "\n",
      "================================================================================\n",
      "plot_distributions.py\n",
      "\tplot_distributions\n",
      "\n",
      "================================================================================\n",
      "get_chromosome_size_from_fasta_gz.py\n",
      "\tget_chromosome_size_from_fasta_gz\n",
      "\n",
      "================================================================================\n",
      "gsea.py\n",
      "\tgsea\n",
      "\n",
      "================================================================================\n",
      "plot_bayesian_nomogram.py\n",
      "\tplot_bayesian_nomogram\n",
      "\n",
      "================================================================================\n",
      "nmf_by_multiple_V_and_H.py\n",
      "\tnmf_by_multiple_V_and_H\n",
      "\n",
      "================================================================================\n",
      "index_gff3_df_by_name.py\n",
      "\tindex_gff3_df_by_name\n",
      "\n",
      "================================================================================\n",
      "bgzip_and_tabix.py\n",
      "\tbgzip_and_tabix\n",
      "\n",
      "================================================================================\n",
      "compute_nd_array_margin_of_error.py\n",
      "\tcompute_nd_array_margin_of_error\n",
      "\n",
      "================================================================================\n",
      "nmf_by_sklearn.py\n",
      "\tnmf_by_sklearn\n",
      "\n",
      "================================================================================\n",
      "__init__.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "_count.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "_make_annotations.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "_make_element_x_dimension.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "plot_pie.py\n",
      "\tplot_pie\n",
      "\n",
      "================================================================================\n",
      "shuffle_each_2d_array_slice.py\n",
      "\tshuffle_each_2d_array_slice\n",
      "\n",
      "================================================================================\n",
      "nmf_consensus_cluster_with_ks.py\n",
      "\tnmf_consensus_cluster_with_ks\n",
      "\n",
      "================================================================================\n",
      "compute_joint_probability.py\n",
      "\tcompute_joint_probability\n",
      "\n",
      "================================================================================\n",
      "compute_mutational_signature_enrichment.py\n",
      "\tcompute_mutational_signature_enrichment\n",
      "\n",
      "================================================================================\n",
      "nd_array_is_sorted.py\n",
      "\tnd_array_is_sorted\n",
      "\n",
      "================================================================================\n",
      "VariantHDF5.py\n",
      "\tVariantHDF5\n",
      "\n",
      "================================================================================\n",
      "conda.py\n",
      "\tinstall_and_activate_conda\n",
      "\tadd_conda_to_path\n",
      "\tconda_is_installed\n",
      "\tget_conda_environments\n",
      "\tget_conda_prefix\n",
      "\n",
      "================================================================================\n",
      "_check_w_or_h.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "GPSMap.py\n",
      "\tGPSMap\n",
      "\n",
      "================================================================================\n",
      "make_case_annotations.py\n",
      "\tmake_case_annotations\n",
      "\n",
      "================================================================================\n",
      "make_summary_match_panel.py\n",
      "\tmake_summary_match_panel\n",
      "\n",
      "================================================================================\n",
      "plot_color_text.py\n",
      "\tplot_color_text\n",
      "\n",
      "================================================================================\n",
      "_style.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "exit_.py\n",
      "\texit_\n",
      "\n",
      "================================================================================\n",
      "plot_heat_map.py\n",
      "\tplot_heat_map\n",
      "\n",
      "================================================================================\n",
      "access_gmt.py\n",
      "\tread_gmts\n",
      "\tread_gmt\n",
      "\twrite_gmt\n",
      "\n",
      "================================================================================\n",
      "_anneal_node_and_element_positions.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "log_nd_array.py\n",
      "\tlog_nd_array\n",
      "\n",
      "================================================================================\n",
      "select_series_low_and_high_index.py\n",
      "\tselect_series_low_and_high_index\n",
      "\n",
      "================================================================================\n",
      "plot_context.py\n",
      "\tplot_context\n",
      "\n",
      "================================================================================\n",
      "make_match_panel.py\n",
      "\tmake_match_panel\n",
      "\n",
      "================================================================================\n",
      "series.py\n",
      "\tcast_series_to_builtins\n",
      "\tmake_membership_df_from_categorical_series\n",
      "\tget_extreme_series_indices\n",
      "\n",
      "================================================================================\n",
      "_update_W_by_multiplicative_update.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "get_gff3_attribute.py\n",
      "\tget_gff3_attribute\n",
      "\n",
      "================================================================================\n",
      "plot_table.py\n",
      "\tplot_table\n",
      "\n",
      "================================================================================\n",
      "mds.py\n",
      "\tmds\n",
      "\n",
      "================================================================================\n",
      "_get_triangulation_edges.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "solve_ax_equal_b.py\n",
      "\tsolve_ax_equal_b\n",
      "\n",
      "================================================================================\n",
      "cluster_2d_array_slices.py\n",
      "\tcluster_2d_array_slices\n",
      "\n",
      "================================================================================\n",
      "process_fasta.py\n",
      "\tfaidx_fasta\n",
      "\n",
      "================================================================================\n",
      "read_where_and_map_column_names.py\n",
      "\tread_where_and_map_column_names\n",
      "\n",
      "================================================================================\n",
      "access_vcf.py\n",
      "\tget_variants_from_vcf_gz\n",
      "\tparse_vcf_row_and_make_variant_dict\n",
      "\tupdate_variant_dict\n",
      "\tcount_gene_impacts_from_variant_dicts\n",
      "\tget_vcf_info\n",
      "\tget_vcf_info_ann\n",
      "\tget_vcf_sample_format\n",
      "\tget_variant_start_and_end_positions\n",
      "\tget_variant_type\n",
      "\tis_inframe\n",
      "\tget_maf_variant_classification\n",
      "\tget_genotype\n",
      "\tget_allelic_frequencies\n",
      "\tget_population_allelic_frequencies\n",
      "\tcount_vcf_gz_rows\n",
      "\n",
      "================================================================================\n",
      "process_sequence.py\n",
      "\ttranscribe_dna_sequence\n",
      "\treverse_transcribe_rna_sequence\n",
      "\treverse_complement_dna_sequence\n",
      "\ttranslate_nucleotide_sequence\n",
      "\tsplit_codons\n",
      "\n",
      "================================================================================\n",
      "_get_coclustering_portion.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "establish_fai_index.py\n",
      "\testablish_fai_index\n",
      "\n",
      "================================================================================\n",
      "path.py\n",
      "\testablish_path\n",
      "\tcopy_path\n",
      "\tremove_paths\n",
      "\tremove_path\n",
      "\tclean_path\n",
      "\tclean_name\n",
      "\n",
      "================================================================================\n",
      "read_mutsignozzlereport2cv.py\n",
      "\tread_mutsignozzlereport2cv\n",
      "\n",
      "================================================================================\n",
      "train_and_classify.py\n",
      "\ttrain_and_classify\n",
      "\n",
      "================================================================================\n",
      "access_gct.py\n",
      "\tread_gct\n",
      "\twrite_gct\n",
      "\n",
      "================================================================================\n",
      "download_clinvar_vcf_gz.py\n",
      "\tdownload_clinvar_vcf_gz\n",
      "\n",
      "================================================================================\n",
      "clip_nd_array_by_standard_deviation.py\n",
      "\tclip_nd_array_by_standard_deviation\n",
      "\n",
      "================================================================================\n",
      "normalize_contig.py\n",
      "\tnormalize_contig\n",
      "\n",
      "================================================================================\n",
      "access_vcf_dict.py\n",
      "\tread_vcf_gz_and_make_vcf_dict\n",
      "\n",
      "================================================================================\n",
      "fit_skew_t_pdfs.py\n",
      "\tfit_skew_t_pdfs\n",
      "\n",
      "================================================================================\n",
      "infer.py\n",
      "\tinfer\n",
      "\n",
      "================================================================================\n",
      "check_nd_array_for_bad.py\n",
      "\tcheck_nd_array_for_bad\n",
      "\n",
      "================================================================================\n",
      "make_categorical_colors.py\n",
      "\tmake_categorical_colors\n",
      "\n",
      "================================================================================\n",
      "_print_and_run_command.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "volume.py\n",
      "\tmount_volume\n",
      "\tunmount_volume\n",
      "\tget_volume_name\n",
      "\tmake_volume_dict\n",
      "\n",
      "================================================================================\n",
      "FeatureHDF5.py\n",
      "\tFeatureHDF5\n",
      "\n",
      "================================================================================\n",
      "get_colormap_colors.py\n",
      "\tget_colormap_colors\n",
      "\n",
      "================================================================================\n",
      "python.py\n",
      "\tget_installed_pip_libraries\n",
      "\tinstall_python_libraries\n",
      "\tget_object_reference\n",
      "\n",
      "================================================================================\n",
      "df.py\n",
      "\tdrop_df_slice_greedily\n",
      "\tdrop_df_slice\n",
      "\tsplit_df\n",
      "\n",
      "================================================================================\n",
      "read_gff3_gz.py\n",
      "\tread_gff3_gz\n",
      "\n",
      "================================================================================\n",
      "compute_information_coefficient.py\n",
      "\tcompute_information_coefficient\n",
      "\n",
      "================================================================================\n",
      "make_reference_genome.py\n",
      "\tmake_reference_genome\n",
      "\n",
      "================================================================================\n",
      "get_intersections_between_2_1d_arrays.py\n",
      "\tget_intersections_between_2_1d_arrays\n",
      "\n",
      "================================================================================\n",
      "select_gene_symbol.py\n",
      "\tselect_gene_symbol\n",
      "\n",
      "================================================================================\n",
      "download_and_parse_geo_data.py\n",
      "\tdownload_and_parse_geo_data\n",
      "\n",
      "================================================================================\n",
      "plot_points.py\n",
      "\tplot_points\n",
      "\n",
      "================================================================================\n",
      "plot_violin_or_box.py\n",
      "\tplot_violin_or_box\n",
      "\n",
      "================================================================================\n",
      "json_.py\n",
      "\tread_json\n",
      "\twrite_json\n",
      "\n",
      "================================================================================\n",
      "style.py\n",
      "\t\n",
      "\n",
      "================================================================================\n",
      "nmf_consensus_cluster.py\n",
      "\tnmf_consensus_cluster\n",
      "\n",
      "================================================================================\n",
      "iterable.py\n",
      "\tgroup_iterable\n",
      "\tflatten_nested_iterable\n",
      "\treplace_bad_objects_in_iterable\n",
      "\tgroup_and_apply_function_on_each_group_in_iterable\n",
      "\tget_unique_iterable_objects_in_order\n",
      "\tmake_object_int_mapping\n",
      "\n",
      "================================================================================\n",
      "cross_validate.py\n",
      "\tcross_validate\n",
      "\n",
      "================================================================================\n",
      "read_correlate_copynumber_vs_mrnaseq.py\n",
      "\tread_correlate_copynumber_vs_mrnaseq\n",
      "\n",
      "================================================================================\n",
      "make_colorscale.py\n",
      "\tmake_colorscale\n",
      "\n",
      "================================================================================\n",
      "math.py\n",
      "\trescale_x_y_coordiantes_in_polar_coordiante\n",
      "\n",
      "================================================================================\n",
      "make_mesh_grid_coordinates_per_axis.py\n",
      "\tmake_mesh_grid_coordinates_per_axis\n",
      "\n",
      "================================================================================\n",
      "hierarchical_consensus_cluster_with_ks.py\n",
      "\thierarchical_consensus_cluster_with_ks\n",
      "\n",
      "================================================================================\n",
      "multiprocess.py\n",
      "\tmultiprocess\n",
      "\n",
      "================================================================================\n",
      "compute_empirical_p_values_and_fdrs.py\n",
      "\tcompute_empirical_p_values_and_fdrs\n",
      "\n",
      "================================================================================\n",
      "_identify_what_to_count.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "_plot_2d.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "access_gps_map.py\n",
      "\tdump_gps_map\n",
      "\tload_gps_map\n",
      "\n",
      "================================================================================\n",
      "process_bam.py\n",
      "\tsort_and_index_bam_using_samtools_sort_and_index\n",
      "\tindex_bam_using_samtools_index\n",
      "\tmark_duplicates_in_bam_using_picard_markduplicates\n",
      "\tcheck_bam_using_samtools_flagstat\n",
      "\tget_variants_from_bam_using_freebayes_and_multiprocess\n",
      "\tget_variants_from_bam_using_freebayes\n",
      "\tget_variants_from_bam_using_strelka\n",
      "\n",
      "================================================================================\n",
      "get_function_name.py\n",
      "\tget_function_name\n",
      "\n",
      "================================================================================\n",
      "compute_correlation_distance.py\n",
      "\tcompute_correlation_distance\n",
      "\n",
      "================================================================================\n",
      "get_1d_array_unique_objects_in_order.py\n",
      "\tget_1d_array_unique_objects_in_order\n",
      "\n",
      "================================================================================\n",
      "summarize_feature_x_sample.py\n",
      "\tsummarize_feature_x_sample\n",
      "\n",
      "================================================================================\n",
      "process_vcf_gz.py\n",
      "\tconcatenate_vcf_gzs_using_bcftools_concat\n",
      "\trename_chromosome_of_vcf_gz_using_bcftools_annotate\n",
      "\tannotate_vcf_gz_using_snpeff\n",
      "\tannotate_vcf_gz_using_bcftools_annotate\n",
      "\tfilter_vcf_gz_using_bcftools_view\n",
      "\n",
      "================================================================================\n",
      "network.py\n",
      "\tdownload\n",
      "\tget_open_port\n",
      "\n",
      "================================================================================\n",
      "compute_bandwidths.py\n",
      "\tcompute_bandwidths\n",
      "\n",
      "================================================================================\n",
      "single_sample_gseas.py\n",
      "\tsingle_sample_gseas\n",
      "\n",
      "================================================================================\n",
      "_match.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "hierarchical_consensus_cluster.py\n",
      "\thierarchical_consensus_cluster\n",
      "\n",
      "================================================================================\n",
      "nmf_by_multiplicative_update.py\n",
      "\tnmf_by_multiplicative_update\n",
      "\n",
      "================================================================================\n",
      "make_comparison_panel.py\n",
      "\tmake_comparison_panel\n",
      "\n",
      "================================================================================\n",
      "_update_H_by_multiplicative_update.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "compute_empirical_p_value.py\n",
      "\tcompute_empirical_p_value\n",
      "\n",
      "================================================================================\n",
      "train_and_regress.py\n",
      "\ttrain_and_regress\n",
      "\n",
      "================================================================================\n",
      "compute_entropy.py\n",
      "\tcompute_entropy\n",
      "\n",
      "================================================================================\n",
      "infer_assuming_independence.py\n",
      "\tinfer_assuming_independence\n",
      "\n",
      "================================================================================\n",
      "correlate.py\n",
      "\tcorrelate\n",
      "\n",
      "================================================================================\n",
      "subprocess_.py\n",
      "\trun_command\n",
      "\trun_command_and_monitor\n",
      "\n",
      "================================================================================\n",
      "_check_node_x_element.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "read_matrix_market.py\n",
      "\tread_matrix_market\n",
      "\n",
      "================================================================================\n",
      "fit_skew_t_pdf.py\n",
      "\tfit_skew_t_pdf\n",
      "\n",
      "================================================================================\n",
      "plot_bubble_map.py\n",
      "\tplot_bubble_map\n",
      "\n",
      "================================================================================\n",
      "solve_for_H.py\n",
      "\tsolve_for_H\n",
      "\n",
      "================================================================================\n",
      "__pycache__\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "make_match_panels.py\n",
      "\tmake_match_panels\n",
      "\n",
      "================================================================================\n",
      "_compute_norm.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "_process_target_or_features_for_plotting.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "str_.py\n",
      "\tcast_str_to_builtins\n",
      "\ttitle_str\n",
      "\tuntitle_str\n",
      "\tsplit_str_ignoring_inside_quotes\n",
      "\tstr_is_version\n",
      "\tmake_file_name_from_str\n",
      "\n",
      "================================================================================\n",
      "read_copynumber_gistic2.py\n",
      "\tread_copynumber_gistic2\n",
      "\n",
      "================================================================================\n",
      "process_feature_x_sample.py\n",
      "\tprocess_feature_x_sample\n",
      "\n",
      "================================================================================\n",
      "_cluster_clustering_x_element_and_compute_ccc.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "access_maf.py\n",
      "\tsplit_maf_by_tumor_sample_barcode\n",
      "\tmake_maf_from_vcf\n",
      "\n",
      "================================================================================\n",
      "_plot_mountain.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "write_dict.py\n",
      "\twrite_dict\n",
      "\n",
      "================================================================================\n",
      "git.py\n",
      "\tcreate_gitkeep\n",
      "\tget_git_versions\n",
      "\tclean_git_url\n",
      "\tin_git_repository\n",
      "\n",
      "================================================================================\n",
      "_compute_context_indices.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "make_context_matrix.py\n",
      "\tmake_context_matrix\n",
      "\n",
      "================================================================================\n",
      "plot_and_save.py\n",
      "\tplot_and_save\n",
      "\n",
      "================================================================================\n",
      "compute_information_distance.py\n",
      "\tcompute_information_distance\n",
      "\n",
      "================================================================================\n",
      "make_random_color.py\n",
      "\tmake_random_color\n",
      "\n",
      "================================================================================\n",
      "simulate_sequences_using_dwgsim.py\n",
      "\tsimulate_sequences_using_dwgsim\n",
      "\n",
      "================================================================================\n",
      "compute_posterior_probability.py\n",
      "\tcompute_posterior_probability\n",
      "\n",
      "================================================================================\n",
      "normalize_nd_array.py\n",
      "\tnormalize_nd_array\n",
      "\n",
      "================================================================================\n",
      "get_sequence_from_fasta_gz.py\n",
      "\tget_sequence_from_fasta_gz\n",
      "\n",
      "================================================================================\n",
      "apply_function_on_2_1d_arrays.py\n",
      "\tapply_function_on_2_1d_arrays\n",
      "\n",
      "================================================================================\n",
      "log.py\n",
      "\tget_now\n",
      "\tinitialize_logger\n",
      "\techo_or_print\n",
      "\tlog_and_return_response\n",
      "\n",
      "================================================================================\n",
      "_plot_gps_map.py\n",
      "\tSkip.\n",
      "\n",
      "================================================================================\n",
      "single_sample_gsea.py\n",
      "\tsingle_sample_gsea\n",
      "\n",
      "================================================================================\n",
      "machine.py\n",
      "\tget_machine\n",
      "\tget_shell_environment\n",
      "\thave_program\n",
      "\tshutdown_machine\n",
      "\treboot_machine\n",
      "\n",
      "================================================================================\n",
      "process_fastq_gz.py\n",
      "\tcheck_fastq_gzs_using_fastqc\n",
      "\ttrim_fastq_gzs_using_skewer\n",
      "\talign_fastq_gzs_using_bwa_mem\n",
      "\talign_fastq_gzs_using_hisat2\n",
      "\tcount_transcripts_using_kallisto_quant\n",
      "\n",
      "================================================================================\n",
      "make_coordinates_for_reflection.py\n",
      "\tmake_coordinates_for_reflection\n",
      "\n",
      "================================================================================\n",
      "Genome.py\n",
      "\tis_valid_vcf_gz\n",
      "\tGenome\n",
      "\n",
      "================================================================================\n",
      "apply_function_on_2_2d_arrays_slices.py\n",
      "\tapply_function_on_2_2d_arrays_slices\n",
      "\n",
      "================================================================================\n",
      "_get_target_grid_indices.py\n",
      "\tSkip.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import_lines = []\n",
    "\n",
    "for py_file_name in os.listdir(\"../ccal\"):\n",
    "\n",
    "    print(\"=\" * 80)\n",
    "\n",
    "    print(py_file_name)\n",
    "\n",
    "    if py_file_name.startswith(\"_\") or not py_file_name.endswith(\".py\"):\n",
    "\n",
    "        print(\"\\tSkip.\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        py_file_path = \"../ccal/{}\".format(py_file_name)\n",
    "\n",
    "        importable_names = get_importable_names_from_py_file_path(py_file_path)\n",
    "\n",
    "        print(\"\\t{}\".format(\"\\n\\t\".join(importable_names)))\n",
    "\n",
    "        if 1 < len(importable_names):\n",
    "\n",
    "            pass\n",
    "#             raise ValueError(\n",
    "#                 \"1 < n importable names ({}): {}\".format(\n",
    "#                     len(importable_names), py_file_name\n",
    "#                 )\n",
    "#             )\n",
    "\n",
    "        import_lines.extend(\n",
    "            [\n",
    "                \"from .{} import {}\".format(\n",
    "                    py_file_name[: -len(\".py\")], importable_name\n",
    "                )\n",
    "                for importable_name in importable_names\n",
    "            ]\n",
    "        )\n",
    "\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../ccal/__init__.py\", mode=\"w\") as init_file:\n",
    "\n",
    "    separator = \"\\n\\n\"\n",
    "\n",
    "    init_file.write(\n",
    "        separator.join(\n",
    "            (\n",
    "                \"from os.path import abspath, dirname\",\n",
    "                'VERSION = \"{}\"'.format(version),\n",
    "                'print(\"CCAL version {} @ {}\".format(VERSION, abspath(__file__)))',\n",
    "                'DATA_DIRECTORY_PATH = \"{}/../data\".format(dirname(__file__))',\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "\n",
    "    init_file.write(separator)\n",
    "\n",
    "    init_file.writelines(\"\\n\".join(import_lines))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='black ../ccal/__init__.py', returncode=0)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subprocess.run(\"isort ../ccal/__init__.py\", shell=True, check=True)\n",
    "\n",
    "subprocess.run(\"black ../ccal/__init__.py\", shell=True, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCAL version 0.9.0 @ /home/kwatme/github/ccal/ccal/__init__.py\n"
     ]
    }
   ],
   "source": [
    "sys.path.insert(0, \"..\")\n",
    "\n",
    "import ccal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "273\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['DATA_DIRECTORY_PATH',\n",
       " 'FeatureHDF5',\n",
       " 'GPSMap',\n",
       " 'Genome',\n",
       " 'VERSION',\n",
       " 'VariantHDF5',\n",
       " '__builtins__',\n",
       " '__cached__',\n",
       " '__doc__',\n",
       " '__file__',\n",
       " '__loader__',\n",
       " '__name__',\n",
       " '__package__',\n",
       " '__path__',\n",
       " '__spec__',\n",
       " '_anneal_node_and_element_positions',\n",
       " '_check_node_x_element',\n",
       " '_check_w_or_h',\n",
       " '_cluster_clustering_x_element_and_compute_ccc',\n",
       " '_compute_context_indices',\n",
       " '_compute_norm',\n",
       " '_count',\n",
       " '_get_coclustering_portion',\n",
       " '_get_target_grid_indices',\n",
       " '_get_triangulation_edges',\n",
       " '_identify_what_to_count',\n",
       " '_make_annotations',\n",
       " '_make_element_x_dimension',\n",
       " '_make_grid_values_and_categorical_labels',\n",
       " '_match',\n",
       " '_plot_2d',\n",
       " '_plot_gps_map',\n",
       " '_plot_mountain',\n",
       " '_print_and_run_command',\n",
       " '_process_target_or_features_for_plotting',\n",
       " '_style',\n",
       " '_update_H_by_multiplicative_update',\n",
       " '_update_W_by_multiplicative_update',\n",
       " 'abspath',\n",
       " 'access_gct',\n",
       " 'access_gmt',\n",
       " 'access_gps_map',\n",
       " 'access_maf',\n",
       " 'access_vcf',\n",
       " 'access_vcf_dict',\n",
       " 'add_conda_to_path',\n",
       " 'align_fastq_gzs_using_bwa_mem',\n",
       " 'align_fastq_gzs_using_hisat2',\n",
       " 'annotate_vcf_gz_using_bcftools_annotate',\n",
       " 'annotate_vcf_gz_using_snpeff',\n",
       " 'apply_function_on_2_1d_arrays',\n",
       " 'apply_function_on_2_2d_arrays_slices',\n",
       " 'bgzip_and_tabix',\n",
       " 'cast_series_to_builtins',\n",
       " 'cast_str_to_builtins',\n",
       " 'check_bam_using_samtools_flagstat',\n",
       " 'check_fastq_gzs_using_fastqc',\n",
       " 'check_nd_array_for_bad',\n",
       " 'clean_git_url',\n",
       " 'clean_name',\n",
       " 'clean_path',\n",
       " 'clip_nd_array_by_standard_deviation',\n",
       " 'cluster_2d_array_slices',\n",
       " 'compression',\n",
       " 'compute_bandwidths',\n",
       " 'compute_context',\n",
       " 'compute_correlation_distance',\n",
       " 'compute_empirical_p_value',\n",
       " 'compute_empirical_p_values_and_fdrs',\n",
       " 'compute_entropy',\n",
       " 'compute_information_coefficient',\n",
       " 'compute_information_distance',\n",
       " 'compute_joint_probability',\n",
       " 'compute_mutational_signature_enrichment',\n",
       " 'compute_nd_array_margin_of_error',\n",
       " 'compute_posterior_probability',\n",
       " 'concatenate_vcf_gzs_using_bcftools_concat',\n",
       " 'conda',\n",
       " 'conda_is_installed',\n",
       " 'copy_path',\n",
       " 'correlate',\n",
       " 'count_gene_impacts_from_variant_dicts',\n",
       " 'count_transcripts_using_kallisto_quant',\n",
       " 'count_vcf_gz_rows',\n",
       " 'create_gitkeep',\n",
       " 'cross_validate',\n",
       " 'df',\n",
       " 'dict_',\n",
       " 'dirname',\n",
       " 'download',\n",
       " 'download_and_parse_geo_data',\n",
       " 'download_clinvar_vcf_gz',\n",
       " 'drop_df_slice',\n",
       " 'drop_df_slice_greedily',\n",
       " 'dump_gps_map',\n",
       " 'echo_or_print',\n",
       " 'establish_fai_index',\n",
       " 'establish_path',\n",
       " 'estimate_kernel_density',\n",
       " 'exit_',\n",
       " 'faidx_fasta',\n",
       " 'filter_vcf_gz_using_bcftools_view',\n",
       " 'fit_skew_t_pdf',\n",
       " 'fit_skew_t_pdfs',\n",
       " 'flatten_nested_iterable',\n",
       " 'get_1d_array_unique_objects_in_order',\n",
       " 'get_allelic_frequencies',\n",
       " 'get_chromosome_size_from_fasta_gz',\n",
       " 'get_colormap_colors',\n",
       " 'get_conda_environments',\n",
       " 'get_conda_prefix',\n",
       " 'get_extreme_series_indices',\n",
       " 'get_function_name',\n",
       " 'get_genotype',\n",
       " 'get_gff3_attribute',\n",
       " 'get_git_versions',\n",
       " 'get_installed_pip_libraries',\n",
       " 'get_intersections_between_2_1d_arrays',\n",
       " 'get_machine',\n",
       " 'get_maf_variant_classification',\n",
       " 'get_now',\n",
       " 'get_object_reference',\n",
       " 'get_open_port',\n",
       " 'get_population_allelic_frequencies',\n",
       " 'get_sequence_from_fasta_gz',\n",
       " 'get_shell_environment',\n",
       " 'get_unique_iterable_objects_in_order',\n",
       " 'get_variant_start_and_end_positions',\n",
       " 'get_variant_type',\n",
       " 'get_variants_from_bam_using_freebayes',\n",
       " 'get_variants_from_bam_using_freebayes_and_multiprocess',\n",
       " 'get_variants_from_bam_using_strelka',\n",
       " 'get_variants_from_vcf_gz',\n",
       " 'get_vcf_info',\n",
       " 'get_vcf_info_ann',\n",
       " 'get_vcf_sample_format',\n",
       " 'get_volume_name',\n",
       " 'git',\n",
       " 'group_and_apply_function_on_each_group_in_iterable',\n",
       " 'group_iterable',\n",
       " 'gsea',\n",
       " 'gzip_compress_file',\n",
       " 'gzip_decompress_and_bgzip_compress_file',\n",
       " 'gzip_decompress_file',\n",
       " 'have_program',\n",
       " 'hierarchical_consensus_cluster',\n",
       " 'hierarchical_consensus_cluster_with_ks',\n",
       " 'in_git_repository',\n",
       " 'index_bam_using_samtools_index',\n",
       " 'index_gff3_df_by_name',\n",
       " 'infer',\n",
       " 'infer_assuming_independence',\n",
       " 'initialize_logger',\n",
       " 'install_and_activate_conda',\n",
       " 'install_python_libraries',\n",
       " 'is_inframe',\n",
       " 'is_valid_vcf_gz',\n",
       " 'iterable',\n",
       " 'json_',\n",
       " 'load_gps_map',\n",
       " 'log',\n",
       " 'log_and_return_response',\n",
       " 'log_nd_array',\n",
       " 'machine',\n",
       " 'make_case_annotations',\n",
       " 'make_categorical_colors',\n",
       " 'make_colorscale',\n",
       " 'make_colorscale_from_colors',\n",
       " 'make_comparison_panel',\n",
       " 'make_context_matrix',\n",
       " 'make_coordinates_for_reflection',\n",
       " 'make_file_name_from_str',\n",
       " 'make_maf_from_vcf',\n",
       " 'make_match_panel',\n",
       " 'make_match_panels',\n",
       " 'make_membership_df_from_categorical_series',\n",
       " 'make_mesh_grid_coordinates_per_axis',\n",
       " 'make_object_int_mapping',\n",
       " 'make_random_color',\n",
       " 'make_reference_genome',\n",
       " 'make_summary_match_panel',\n",
       " 'make_volume_dict',\n",
       " 'mark_duplicates_in_bam_using_picard_markduplicates',\n",
       " 'math',\n",
       " 'mds',\n",
       " 'merge_dicts_with_callable',\n",
       " 'mount_volume',\n",
       " 'multiprocess',\n",
       " 'nd_array_is_sorted',\n",
       " 'network',\n",
       " 'nmf_by_multiple_V_and_H',\n",
       " 'nmf_by_multiplicative_update',\n",
       " 'nmf_by_sklearn',\n",
       " 'nmf_consensus_cluster',\n",
       " 'nmf_consensus_cluster_with_ks',\n",
       " 'normalize_contig',\n",
       " 'normalize_nd_array',\n",
       " 'parse_vcf_row_and_make_variant_dict',\n",
       " 'path',\n",
       " 'plot_and_save',\n",
       " 'plot_bar',\n",
       " 'plot_bayesian_nomogram',\n",
       " 'plot_bubble_map',\n",
       " 'plot_color_text',\n",
       " 'plot_context',\n",
       " 'plot_distributions',\n",
       " 'plot_heat_map',\n",
       " 'plot_pie',\n",
       " 'plot_points',\n",
       " 'plot_table',\n",
       " 'plot_violin_or_box',\n",
       " 'process_bam',\n",
       " 'process_fasta',\n",
       " 'process_fastq_gz',\n",
       " 'process_feature_x_sample',\n",
       " 'process_sequence',\n",
       " 'process_vcf_gz',\n",
       " 'python',\n",
       " 'read_copynumber_gistic2',\n",
       " 'read_correlate_copynumber_vs_mrnaseq',\n",
       " 'read_gct',\n",
       " 'read_gff3_gz',\n",
       " 'read_gmt',\n",
       " 'read_gmts',\n",
       " 'read_json',\n",
       " 'read_matrix_market',\n",
       " 'read_mutsignozzlereport2cv',\n",
       " 'read_vcf_gz_and_make_vcf_dict',\n",
       " 'read_where_and_map_column_names',\n",
       " 'reboot_machine',\n",
       " 'remove_path',\n",
       " 'remove_paths',\n",
       " 'rename_chromosome_of_vcf_gz_using_bcftools_annotate',\n",
       " 'replace_bad_objects_in_iterable',\n",
       " 'rescale_x_y_coordiantes_in_polar_coordiante',\n",
       " 'reverse_complement_dna_sequence',\n",
       " 'reverse_transcribe_rna_sequence',\n",
       " 'run_command',\n",
       " 'run_command_and_monitor',\n",
       " 'select_gene_symbol',\n",
       " 'select_series_low_and_high_index',\n",
       " 'series',\n",
       " 'shuffle_each_2d_array_slice',\n",
       " 'shutdown_machine',\n",
       " 'simulate_sequences_using_dwgsim',\n",
       " 'single_sample_gsea',\n",
       " 'single_sample_gseas',\n",
       " 'solve_ax_equal_b',\n",
       " 'solve_for_H',\n",
       " 'sort_and_index_bam_using_samtools_sort_and_index',\n",
       " 'split_codons',\n",
       " 'split_df',\n",
       " 'split_maf_by_tumor_sample_barcode',\n",
       " 'split_str_ignoring_inside_quotes',\n",
       " 'str_',\n",
       " 'str_is_version',\n",
       " 'style',\n",
       " 'subprocess_',\n",
       " 'summarize_feature_x_sample',\n",
       " 'title_str',\n",
       " 'train_and_classify',\n",
       " 'train_and_regress',\n",
       " 'transcribe_dna_sequence',\n",
       " 'translate_nucleotide_sequence',\n",
       " 'trim_fastq_gzs_using_skewer',\n",
       " 'unmount_volume',\n",
       " 'untitle_str',\n",
       " 'update_variant_dict',\n",
       " 'volume',\n",
       " 'write_dict',\n",
       " 'write_gct',\n",
       " 'write_gmt',\n",
       " 'write_json']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(dir(ccal)))\n",
    "\n",
    "dir(ccal)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
