{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This is a environment to test, prototype, and develop CCBA Library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<12:28:43.267826> Checking dependencies ...\n",
      "<12:28:43.270066> Using the following packages:\n",
      "<12:28:43.271697> \tmatplotlib v1.5.1\n",
      "<12:28:43.271714> \tnumpy v1.10.4\n",
      "<12:28:43.271723> \tpandas v0.18.0\n",
      "<12:28:43.271743> \trpy2 v2.7.9\n",
      "<12:28:43.271751> \tscikit-learn v0.17.1\n",
      "<12:28:43.271758> \tscipy v0.17.0\n",
      "<12:28:43.271765> \tseaborn v0.7.0\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "sys.path.insert(0, '/Users/Kwat/binf/ccba')\n",
    "import ccba_lib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>40</th>\n",
       "      <th>41</th>\n",
       "      <th>42</th>\n",
       "      <th>43</th>\n",
       "      <th>44</th>\n",
       "      <th>45</th>\n",
       "      <th>46</th>\n",
       "      <th>47</th>\n",
       "      <th>48</th>\n",
       "      <th>49</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GENE 0</th>\n",
       "      <td>0.051271</td>\n",
       "      <td>0.921401</td>\n",
       "      <td>0.191297</td>\n",
       "      <td>0.985135</td>\n",
       "      <td>0.595941</td>\n",
       "      <td>0.311613</td>\n",
       "      <td>0.628693</td>\n",
       "      <td>0.264480</td>\n",
       "      <td>0.930568</td>\n",
       "      <td>0.637047</td>\n",
       "      <td>...</td>\n",
       "      <td>0.275756</td>\n",
       "      <td>0.596258</td>\n",
       "      <td>0.801343</td>\n",
       "      <td>0.520996</td>\n",
       "      <td>0.902098</td>\n",
       "      <td>0.858173</td>\n",
       "      <td>0.073735</td>\n",
       "      <td>0.320174</td>\n",
       "      <td>0.951701</td>\n",
       "      <td>0.056643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE 1</th>\n",
       "      <td>0.174270</td>\n",
       "      <td>0.417102</td>\n",
       "      <td>0.968662</td>\n",
       "      <td>0.729003</td>\n",
       "      <td>0.153189</td>\n",
       "      <td>0.649962</td>\n",
       "      <td>0.197777</td>\n",
       "      <td>0.182436</td>\n",
       "      <td>0.584316</td>\n",
       "      <td>0.917793</td>\n",
       "      <td>...</td>\n",
       "      <td>0.190018</td>\n",
       "      <td>0.867556</td>\n",
       "      <td>0.265831</td>\n",
       "      <td>0.291223</td>\n",
       "      <td>0.474530</td>\n",
       "      <td>0.684393</td>\n",
       "      <td>0.141573</td>\n",
       "      <td>0.689306</td>\n",
       "      <td>0.106853</td>\n",
       "      <td>0.441907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE 2</th>\n",
       "      <td>0.989138</td>\n",
       "      <td>0.756691</td>\n",
       "      <td>0.207953</td>\n",
       "      <td>0.835845</td>\n",
       "      <td>0.414697</td>\n",
       "      <td>0.392975</td>\n",
       "      <td>0.760158</td>\n",
       "      <td>0.177299</td>\n",
       "      <td>0.964068</td>\n",
       "      <td>0.056773</td>\n",
       "      <td>...</td>\n",
       "      <td>0.711543</td>\n",
       "      <td>0.050046</td>\n",
       "      <td>0.256184</td>\n",
       "      <td>0.380116</td>\n",
       "      <td>0.928233</td>\n",
       "      <td>0.665801</td>\n",
       "      <td>0.865007</td>\n",
       "      <td>0.318722</td>\n",
       "      <td>0.627798</td>\n",
       "      <td>0.911761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE 3</th>\n",
       "      <td>0.953946</td>\n",
       "      <td>0.143804</td>\n",
       "      <td>0.177641</td>\n",
       "      <td>0.765037</td>\n",
       "      <td>0.609406</td>\n",
       "      <td>0.651566</td>\n",
       "      <td>0.045911</td>\n",
       "      <td>0.516741</td>\n",
       "      <td>0.890781</td>\n",
       "      <td>0.764801</td>\n",
       "      <td>...</td>\n",
       "      <td>0.272247</td>\n",
       "      <td>0.193629</td>\n",
       "      <td>0.542403</td>\n",
       "      <td>0.073893</td>\n",
       "      <td>0.758281</td>\n",
       "      <td>0.154146</td>\n",
       "      <td>0.798682</td>\n",
       "      <td>0.056344</td>\n",
       "      <td>0.313064</td>\n",
       "      <td>0.669908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE 4</th>\n",
       "      <td>0.578882</td>\n",
       "      <td>0.014461</td>\n",
       "      <td>0.473863</td>\n",
       "      <td>0.967510</td>\n",
       "      <td>0.457812</td>\n",
       "      <td>0.189865</td>\n",
       "      <td>0.661590</td>\n",
       "      <td>0.420299</td>\n",
       "      <td>0.197117</td>\n",
       "      <td>0.198698</td>\n",
       "      <td>...</td>\n",
       "      <td>0.792833</td>\n",
       "      <td>0.895478</td>\n",
       "      <td>0.550158</td>\n",
       "      <td>0.909920</td>\n",
       "      <td>0.259950</td>\n",
       "      <td>0.199784</td>\n",
       "      <td>0.828053</td>\n",
       "      <td>0.852697</td>\n",
       "      <td>0.897329</td>\n",
       "      <td>0.048526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE 5</th>\n",
       "      <td>0.805414</td>\n",
       "      <td>0.057471</td>\n",
       "      <td>0.027730</td>\n",
       "      <td>0.741032</td>\n",
       "      <td>0.342279</td>\n",
       "      <td>0.014748</td>\n",
       "      <td>0.631572</td>\n",
       "      <td>0.958271</td>\n",
       "      <td>0.049869</td>\n",
       "      <td>0.793396</td>\n",
       "      <td>...</td>\n",
       "      <td>0.328902</td>\n",
       "      <td>0.311373</td>\n",
       "      <td>0.358680</td>\n",
       "      <td>0.880204</td>\n",
       "      <td>0.253496</td>\n",
       "      <td>0.245316</td>\n",
       "      <td>0.702735</td>\n",
       "      <td>0.674238</td>\n",
       "      <td>0.859618</td>\n",
       "      <td>0.111661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE 6</th>\n",
       "      <td>0.442554</td>\n",
       "      <td>0.878732</td>\n",
       "      <td>0.873483</td>\n",
       "      <td>0.271223</td>\n",
       "      <td>0.194238</td>\n",
       "      <td>0.513945</td>\n",
       "      <td>0.207285</td>\n",
       "      <td>0.497829</td>\n",
       "      <td>0.626682</td>\n",
       "      <td>0.238229</td>\n",
       "      <td>...</td>\n",
       "      <td>0.959766</td>\n",
       "      <td>0.985660</td>\n",
       "      <td>0.844935</td>\n",
       "      <td>0.886476</td>\n",
       "      <td>0.418662</td>\n",
       "      <td>0.868782</td>\n",
       "      <td>0.454037</td>\n",
       "      <td>0.681551</td>\n",
       "      <td>0.886062</td>\n",
       "      <td>0.080040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE 7</th>\n",
       "      <td>0.954337</td>\n",
       "      <td>0.328683</td>\n",
       "      <td>0.609777</td>\n",
       "      <td>0.344893</td>\n",
       "      <td>0.406600</td>\n",
       "      <td>0.219783</td>\n",
       "      <td>0.533871</td>\n",
       "      <td>0.476177</td>\n",
       "      <td>0.568830</td>\n",
       "      <td>0.521045</td>\n",
       "      <td>...</td>\n",
       "      <td>0.377055</td>\n",
       "      <td>0.909171</td>\n",
       "      <td>0.783395</td>\n",
       "      <td>0.565553</td>\n",
       "      <td>0.044490</td>\n",
       "      <td>0.514733</td>\n",
       "      <td>0.265224</td>\n",
       "      <td>0.412123</td>\n",
       "      <td>0.961553</td>\n",
       "      <td>0.362470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE 8</th>\n",
       "      <td>0.518900</td>\n",
       "      <td>0.424272</td>\n",
       "      <td>0.958373</td>\n",
       "      <td>0.105200</td>\n",
       "      <td>0.088094</td>\n",
       "      <td>0.441740</td>\n",
       "      <td>0.264603</td>\n",
       "      <td>0.184418</td>\n",
       "      <td>0.907635</td>\n",
       "      <td>0.796598</td>\n",
       "      <td>...</td>\n",
       "      <td>0.406262</td>\n",
       "      <td>0.623524</td>\n",
       "      <td>0.401726</td>\n",
       "      <td>0.516787</td>\n",
       "      <td>0.924307</td>\n",
       "      <td>0.155526</td>\n",
       "      <td>0.064867</td>\n",
       "      <td>0.225622</td>\n",
       "      <td>0.000834</td>\n",
       "      <td>0.326111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE 9</th>\n",
       "      <td>0.157120</td>\n",
       "      <td>0.086947</td>\n",
       "      <td>0.970490</td>\n",
       "      <td>0.590287</td>\n",
       "      <td>0.516507</td>\n",
       "      <td>0.176259</td>\n",
       "      <td>0.180591</td>\n",
       "      <td>0.604037</td>\n",
       "      <td>0.731047</td>\n",
       "      <td>0.118600</td>\n",
       "      <td>...</td>\n",
       "      <td>0.203036</td>\n",
       "      <td>0.976647</td>\n",
       "      <td>0.303970</td>\n",
       "      <td>0.682681</td>\n",
       "      <td>0.588631</td>\n",
       "      <td>0.274092</td>\n",
       "      <td>0.584769</td>\n",
       "      <td>0.355322</td>\n",
       "      <td>0.287011</td>\n",
       "      <td>0.787916</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows Ã— 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              0         1         2         3         4         5         6   \\\n",
       "GENE 0  0.051271  0.921401  0.191297  0.985135  0.595941  0.311613  0.628693   \n",
       "GENE 1  0.174270  0.417102  0.968662  0.729003  0.153189  0.649962  0.197777   \n",
       "GENE 2  0.989138  0.756691  0.207953  0.835845  0.414697  0.392975  0.760158   \n",
       "GENE 3  0.953946  0.143804  0.177641  0.765037  0.609406  0.651566  0.045911   \n",
       "GENE 4  0.578882  0.014461  0.473863  0.967510  0.457812  0.189865  0.661590   \n",
       "GENE 5  0.805414  0.057471  0.027730  0.741032  0.342279  0.014748  0.631572   \n",
       "GENE 6  0.442554  0.878732  0.873483  0.271223  0.194238  0.513945  0.207285   \n",
       "GENE 7  0.954337  0.328683  0.609777  0.344893  0.406600  0.219783  0.533871   \n",
       "GENE 8  0.518900  0.424272  0.958373  0.105200  0.088094  0.441740  0.264603   \n",
       "GENE 9  0.157120  0.086947  0.970490  0.590287  0.516507  0.176259  0.180591   \n",
       "\n",
       "              7         8         9     ...           40        41        42  \\\n",
       "GENE 0  0.264480  0.930568  0.637047    ...     0.275756  0.596258  0.801343   \n",
       "GENE 1  0.182436  0.584316  0.917793    ...     0.190018  0.867556  0.265831   \n",
       "GENE 2  0.177299  0.964068  0.056773    ...     0.711543  0.050046  0.256184   \n",
       "GENE 3  0.516741  0.890781  0.764801    ...     0.272247  0.193629  0.542403   \n",
       "GENE 4  0.420299  0.197117  0.198698    ...     0.792833  0.895478  0.550158   \n",
       "GENE 5  0.958271  0.049869  0.793396    ...     0.328902  0.311373  0.358680   \n",
       "GENE 6  0.497829  0.626682  0.238229    ...     0.959766  0.985660  0.844935   \n",
       "GENE 7  0.476177  0.568830  0.521045    ...     0.377055  0.909171  0.783395   \n",
       "GENE 8  0.184418  0.907635  0.796598    ...     0.406262  0.623524  0.401726   \n",
       "GENE 9  0.604037  0.731047  0.118600    ...     0.203036  0.976647  0.303970   \n",
       "\n",
       "              43        44        45        46        47        48        49  \n",
       "GENE 0  0.520996  0.902098  0.858173  0.073735  0.320174  0.951701  0.056643  \n",
       "GENE 1  0.291223  0.474530  0.684393  0.141573  0.689306  0.106853  0.441907  \n",
       "GENE 2  0.380116  0.928233  0.665801  0.865007  0.318722  0.627798  0.911761  \n",
       "GENE 3  0.073893  0.758281  0.154146  0.798682  0.056344  0.313064  0.669908  \n",
       "GENE 4  0.909920  0.259950  0.199784  0.828053  0.852697  0.897329  0.048526  \n",
       "GENE 5  0.880204  0.253496  0.245316  0.702735  0.674238  0.859618  0.111661  \n",
       "GENE 6  0.886476  0.418662  0.868782  0.454037  0.681551  0.886062  0.080040  \n",
       "GENE 7  0.565553  0.044490  0.514733  0.265224  0.412123  0.961553  0.362470  \n",
       "GENE 8  0.516787  0.924307  0.155526  0.064867  0.225622  0.000834  0.326111  \n",
       "GENE 9  0.682681  0.588631  0.274092  0.584769  0.355322  0.287011  0.787916  \n",
       "\n",
       "[10 rows x 50 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Make test data\n",
    "nrow = 10\n",
    "ncol = 50\n",
    "ref = np.random.rand(ncol)\n",
    "df = pd.DataFrame(np.random.random((nrow, ncol)))\n",
    "df.index = ['GENE {}'.format(i) for i in range(nrow)]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<12:28:55.228069> Perfomring NMF with k=3 ...\n"
     ]
    }
   ],
   "source": [
    "nmf_result = ccba_lib.analyze.nmf(df, [3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# make_heatmap_panel is in the ccba/library/ccba.py\n",
    "# make_heatmap_panel calls plot_heatmap_panel, which is in the ccba/library/visualize.py\n",
    "# make_heatmap_panel computes IC of each df's row and the ref (target).\n",
    "# In the future, we can add other computation within this function so it can compute P-Val and FDR as well.\n",
    "make_heatmap_panel(df, ref, ['IC', 'P-Val'])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# TODO: learn what to exclude\n",
    "# Filter x and y\n",
    "# TODO: understanding why intersect\n",
    "# Get the intersection between x and y\n",
    "if intersection:\n",
    "    # Choose the rank of an SVD approximation via cross validation for x and y\n",
    "    rank_x =\n",
    "    rank_y =\n",
    "    \n",
    "    # Compute the |correlation| of x and y\n",
    "    cor = abs()\n",
    "    \n",
    "    # TODO: understand where the numbers come from\n",
    "    rank_x_XXX =\n",
    "    rank_y_XXX =\n",
    "\n",
    "    # TODO: \n",
    "    # Compute 2-dimensional kernel density estimation with an axis-aligned bivariate normal kernel, evaluated on a square grid.\n",
    "    kernel_x =\n",
    "    kernel_y =\n",
    "    kernel_density =\n",
    "    \n",
    "    # TODO: understand math\n",
    "    dx = \n",
    "    dy = \n",
    "    prob_xy = \n",
    "    prob_x = \n",
    "    prob_y = \n",
    "    entropy_xy = \n",
    "    entropy_x = \n",
    "    entropy_y = \n",
    "    prob_x = \n",
    "    prob_y = \n",
    "    mutual_information = \n",
    "    information_coefficient = \n",
    "else:\n",
    "    information_coefficient = 0\n",
    "    \n",
    "print(information_coefficient) #return information_coefficient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import cProfile\n",
    "#import timeit\n",
    "import time\n",
    "from library.ccba import *\n",
    "from scipy.stats import pearsonr\n",
    "import seaborn as sns\n",
    "%matplotlib inline\n",
    "\n",
    "def runtime(function, n_range, verbose=False):\n",
    "    ns = []\n",
    "    runtimes = []\n",
    "    for i in n_range:\n",
    "        n = 10 * (i + 1)\n",
    "        if verbose:\n",
    "            print(n)\n",
    "        x = np.random.rand(n)\n",
    "        y = np.random.rand(n)\n",
    "        t0 = time.time()\n",
    "\n",
    "        function(x, y)\n",
    "\n",
    "        runtime = time.time() - t0\n",
    "        ns.append(n)\n",
    "        runtimes.append(runtime)\n",
    "        #cProfile.run(func)\n",
    "    return ns, runtimes\n",
    "\n",
    "ns, runtimes = runtime(information_coefficient, n_range=list(range(1000)))\n",
    "ax = sns.pointplot(x=ns, y=runtimes)\n",
    "\n",
    "ns, runtimes = runtime(pearsonr, n_range=list(range(1000)), verbose=False)\n",
    "ax = sns.pointplot(x=ns, y=runtimes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Feature selection test with real data for IC, CMI diff, and CMI ratio association metrics\n",
    "## Ranking 60 genes + target profile with respect to the activation profile of KRAS component 6 (BRAF/MAPK)\n",
    "\n",
    "def plot_heatmap_panel_v2(target, df, row_annot, title, t_type=\"auto\", sort_target=True, class_labels=None, cbar=False):\n",
    "    \"\"\"\n",
    "    Plot horizonzal heatmap panels.s\n",
    "    \"\"\"     \n",
    "    ncol = df.shape[1]\n",
    "    nrow = df.shape[0]\n",
    "    df2 = pd.DataFrame(np.zeros((nrow, ncol)))\n",
    "    CMAP_CONTINUOUS =  mpl.cm.bwr \n",
    "    CMAP_BINARY = sns.light_palette(\"black\", n_colors=128, as_cmap=True)\n",
    "    CMAP_CATEGORICAL = mpl.cm.Set2 \n",
    "    f_width = np.where(ncol/5 < 5, 5, np.where(ncol/5 > 7, 7, ncol/5))\n",
    "    f_height = np.where(nrow < 25, nrow/1.5, nrow/2.5)\n",
    "    figure_size = (f_width, f_height)\n",
    "\n",
    "    # Target pre-processing\n",
    "    if t_type == \"binary\" or (t_type == \"auto\" and len(np.unique(target.iloc[0,:]))) == 2:\n",
    "        cmap_target = CMAP_BINARY\n",
    "        (t_vmin, t_vmax) = (0, 1)\n",
    "    elif t_type == \"categ\" or (t_type == \"auto\" and len(np.unique(target.iloc[0,:])) > 2 and \n",
    "                               len(np.unique(target.iloc[0,:])) < ncol): \n",
    "        cmap_target = CMAP_CATEGORICAL\n",
    "        (t_vmin, t_vmax) = (0, len(np.unique(target.iloc[0:]) - 1))   \n",
    "    else:  # t_type = \"cont\"      \n",
    "        cmap_target = CMAP_CONTINUOUS\n",
    "        target.iloc[0,:] = (target.iloc[0,:] - np.mean(target.iloc[0,:]))/np.std(target.iloc[0,:])\n",
    "        (t_vmin, t_vmax) = (-2.5, 2.5)\n",
    "        means = df.mean(1, skipna=True)\n",
    "        stds = df.std(1, skipna=True) \n",
    "\n",
    "    # Feature pre-processing\n",
    "    row_names = df.index \n",
    "    max_len_row = max([ len(x) for x in row_names])\n",
    "    if len(np.unique(df)) == 2:\n",
    "        cmap_features = CMAP_BINARY\n",
    "        (f_vmin, f_vmax) = (0, 1)\n",
    "        df = df + 0.25\n",
    "    else:\n",
    "        cmap_features = CMAP_CONTINUOUS\n",
    "        (f_vmin, f_vmax) = (-2.5, 2.5)\n",
    "        means = df.mean(1, skipna=True)\n",
    "        stds = df.std(1, skipna=True)\n",
    "        for i in range(nrow):\n",
    "            for j in range(ncol):\n",
    "                df.values[i, j] = (df.values[i, j] - means[i])/stds[i]\n",
    "         \n",
    "    if sort_target == True:\n",
    "        t_order = list(reversed(np.argsort(target.values[0,:], kind='quicksort')))\n",
    "        target = target.reindex_axis(target.columns[t_order], axis=1) \n",
    "        df2 = df.reindex_axis(df.columns[t_order], axis=1)\n",
    "  \n",
    "        # Computer locations for class labels (for binary or categorical)\n",
    "        if (t_type == \"binary\" or t_type == \"categ\") and class_labels != None:\n",
    "            boundaries = np.zeros(len(np.unique(target.iloc[0,:])))\n",
    "            locs_labels = np.zeros(len(np.unique(target.iloc[0,:])))\n",
    "            k = 0\n",
    "            for i in range(1, ncol):\n",
    "                if target.iloc[0, i] != target.iloc[0, i-1]:\n",
    "                    boundaries[k] = i \n",
    "                    k = k + 1\n",
    "            boundaries[len(boundaries) - 1] = ncol\n",
    "            locs_labels[0] = boundaries[0]/2\n",
    "            for k in range(1, len(locs_labels)):\n",
    "                locs_labels[k] = boundaries[k] - (boundaries[k] - boundaries[k - 1])/2.0  \n",
    "        \n",
    "    ## Initialize figure\n",
    "    fig = plt.figure(figsize=figure_size)\n",
    "    \n",
    "    ax1 = plt.subplot2grid((nrow + 1, 1), (0, 0))\n",
    "\n",
    "    sns.heatmap(target, vmin=t_vmin, vmax=t_vmax, robust=True, center=None, mask=None,\n",
    "            square=False, cmap=cmap_target, linewidth=0.0, linecolor='b',\n",
    "            annot=False, fmt=None, annot_kws={}, xticklabels=False, \n",
    "            yticklabels=['  '], cbar=False)\n",
    "    ax1.text(-0.1, 0.2, target.index[0], fontsize=13, horizontalalignment='right', fontweight='bold')\n",
    "    ax1.text(ncol/2, 1.6, title, fontsize=16, horizontalalignment='center', fontweight='bold')\n",
    "    if (t_type == \"binary\" or t_type == \"categ\") and class_labels != None:\n",
    "        for k in range(len(locs_labels)):\n",
    "            ax1.text(locs_labels[k], 0.25, class_labels[k], fontsize=13, horizontalalignment='center', fontweight='bold')\n",
    "   \n",
    "    ax2 = plt.subplot2grid((nrow + 1, 1), (0, 1), rowspan=nrow)\n",
    "    sns.heatmap(df2, vmin=f_vmin, vmax=f_vmax, robust=True, center=None, mask=None,\n",
    "            square=False, cmap=cmap_features, linewidth=0.0, linecolor='b',\n",
    "            annot=False, fmt=None, annot_kws={}, xticklabels=False,                \n",
    "            yticklabels=[' ' for i in range(nrow)], cbar=cbar, cbar_kws={\"orientation\": \"horizontal\"})  \n",
    "    for i in range(nrow):\n",
    "        ax2.text(-0.1, nrow - i - 1 + 0.3, row_names[i], fontsize=13, horizontalalignment='right', fontweight='bold')\n",
    "        ax2.text(ncol + 0.1, nrow - i - 1 + 0.3, row_annot[i], fontsize=13, fontweight='bold')\n",
    "    ax2.text(ncol + 1, nrow + 0.3, row_annot.name, fontsize=13, fontweight='bold')        \n",
    "    \n",
    "    fig.tight_layout()    \n",
    "    plt.show(fig) \n",
    "    \n",
    "    return\n",
    "\n",
    "## Feature selection using Information Coefficient\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "gct1 = read_gct('/Users/tamayo/UCSD_2015/signatures/paper2/analysis/top_hits.gct', fill_na=0)\n",
    "gct2 = read_gct('/Users/tamayo/UCSD_2015/signatures/Onco-GPS/CCLE_RNAseq_KRAS_SALE_Late.No_HAEM.k_9.H.v1.gct', fill_na=0)\n",
    "\n",
    "common_cols = list(set(gct1.columns) & set(gct2.columns))\n",
    "A = gct1[common_cols]\n",
    "B = gct2[common_cols]\n",
    "target = pd.DataFrame(B.loc['KRAS_SALE_Late_Comp_C6_9'])\n",
    "target = target.T\n",
    "\n",
    "IC = np.zeros(A.shape[0])\n",
    "for i in range(A.shape[0]):\n",
    "    IC[i] = round(compute_information_coefficient(x=target.iloc[0,:], y=A.iloc[i,:]), 3)  \n",
    "                \n",
    "IC_order = list(np.argsort(IC, kind='quicksort'))\n",
    "IC = IC[IC_order]\n",
    "A_sorted = A.reindex(A.index[IC_order]) \n",
    "\n",
    "end_time = time.time() \n",
    "print(\"Feature selection time %s seconds \" % (end_time - start_time))\n",
    "\n",
    "IC_val = pd.Series(IC[:A.shape[0]], name = \"IC\")\n",
    "features = A_sorted.head(A.shape[0])\n",
    "plot_heatmap_panel_v2(target, features, IC_val, \"KRAS component 6 (BRAF/MAPK) co-essential genes\")\n",
    "\n",
    "## Feature selection using CMI diff\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "gct1 = read_gct('/Users/tamayo/UCSD_2015/signatures/paper2/analysis/top_hits.gct', fill_na=0)\n",
    "gct2 = read_gct('/Users/tamayo/UCSD_2015/signatures/Onco-GPS/CCLE_RNAseq_KRAS_SALE_Late.No_HAEM.k_9.H.v1.gct', fill_na=0)\n",
    "\n",
    "common_cols = list(set(gct1.columns) & set(gct2.columns))\n",
    "A = gct1[common_cols]\n",
    "B = gct2[common_cols]\n",
    "target = pd.DataFrame(B.loc['KRAS_SALE_Late_Comp_C6_9'])\n",
    "target = target.T\n",
    "\n",
    "CMI_diff_vals = np.zeros(A.shape[0])\n",
    "for i in range(A.shape[0]):\n",
    "    CMI_diff_vals[i] = round(CMI_diff(x=target.iloc[0,:], y=A.iloc[i,:]), 3)  \n",
    "                \n",
    "CMI_order = list(np.argsort(CMI_diff_vals, kind='quicksort'))\n",
    "CMI_diff_vals = CMI_diff_vals[CMI_order]\n",
    "A_sorted = A.reindex(A.index[CMI_order]) \n",
    "\n",
    "end_time = time.time() \n",
    "print(\"Feature selection time %s seconds \" % (end_time - start_time))\n",
    "\n",
    "CMI_val = pd.Series(CMI_diff_vals[:A.shape[0]], name = \"CMI diff\")\n",
    "features = A_sorted.head(A.shape[0])\n",
    "plot_heatmap_panel_v2(target, features, CMI_val, \"KRAS component 6 (BRAF/MAPK) co-essential genes\")\n",
    "\n",
    "## Feature selection using CMI ratio\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "\n",
    "gct1 = read_gct('/Users/tamayo/UCSD_2015/signatures/paper2/analysis/top_hits.gct', fill_na=0)\n",
    "gct2 = read_gct('/Users/tamayo/UCSD_2015/signatures/Onco-GPS/CCLE_RNAseq_KRAS_SALE_Late.No_HAEM.k_9.H.v1.gct', fill_na=0)\n",
    "\n",
    "common_cols = list(set(gct1.columns) & set(gct2.columns))\n",
    "A = gct1[common_cols]\n",
    "B = gct2[common_cols]\n",
    "target = pd.DataFrame(B.loc['KRAS_SALE_Late_Comp_C6_9'])\n",
    "target = target.T\n",
    "\n",
    "CMI_ratio_vals = np.zeros(A.shape[0])\n",
    "for i in range(A.shape[0]):\n",
    "    CMI_ratio_vals[i] = round(CMI_ratio(x=target.iloc[0,:], y=A.iloc[i,:]), 3)  \n",
    "                \n",
    "CMI_order = list(np.argsort(CMI_ratio_vals, kind='quicksort'))\n",
    "CMI_ratio_vals = CMI_ratio_vals[CMI_order]\n",
    "A_sorted = A.reindex(A.index[CMI_order]) \n",
    "\n",
    "end_time = time.time() \n",
    "print(\"Feature selection time %s seconds \" % (end_time - start_time))\n",
    "\n",
    "CMI_val = pd.Series(CMI_ratio_vals[:A.shape[0]], name = \"CMI ratio\")\n",
    "features = A_sorted.head(A.shape[0])\n",
    "\n",
    "plot_heatmap_panel_v2(target, features, CMI_val, \"KRAS component 6 (BRAF/MAPK) co-essential genes\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
